<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 7.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="白马？定叫他有来无回！">
<meta property="og:type" content="website">
<meta property="og:title" content="WeFun blog">
<meta property="og:url" content="http://example.com/index.html">
<meta property="og:site_name" content="WeFun blog">
<meta property="og:description" content="白马？定叫他有来无回！">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="Plank">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'zh-CN'
  };
</script>

  <title>WeFun blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">WeFun blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>归档</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2025/04/18/circos/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="Plank">
      <meta itemprop="description" content="白马？定叫他有来无回！">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="WeFun blog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2025/04/18/circos/" class="post-title-link" itemprop="url">circos图的绘制</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>
              

              <time title="创建时间：2025-04-18 20:02:45 / 修改时间：20:16:49" itemprop="dateCreated datePublished" datetime="2025-04-18T20:02:45+08:00">2025-04-18</time>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <pre><code>绘制circos图可以使用circos软件在服务器中运行，也可以使用TBtools，两者出图没有本质的区别，在此介绍对小白更友好，操作更简单的TBtools，但是一些文件的获得依旧需要依靠服务器.  
教程来源主要为：https://www.yuque.com/cjchen/hirv8i/b88df7179632c0ee5f86a4b74a58f933  
此外含有大量个人经验，仅看本教程完全可以完成全部工作
</code></pre>
<h2 id="1-准备文件"><a href="#1-准备文件" class="headerlink" title="1.准备文件"></a>1.准备文件</h2><p>原始文件为chr.fasta（如果有除了染色体之外的序列，需要全部删去），all_chr.gff（重复序列注释文件）以及Msin.chr.gff（基因组注释文件）<br>然后准备文件（*表示必须）<br>1）*染色体长度文件<br>2）基因密度文件<br>3）TE密度文件<br>4）COPIA密度文件<br>5）GYPSY密度文件<br>6）GC含量密度文件<br>7）共线性文件  </p>
<p><strong>1）染色体长度文件</strong></p>
<p>TBtools-&gt;sequence toolkits-&gt;fasta tools-&gt;fasta stater<br>第一栏拖入chr.fasta，设置输出文件路径（需指定文件名），运行，提取出第一列及第四列中的数据，即得到染色体长度文件<br>也可以使用samtools和bedtools来完成，后面的文件也都需要这两个软件，图省事可以使用conda安装  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget https://github.com/arq5x/bedtools2/archive/v2.30.0.tar.gz</span><br><span class="line">tar -zxvf v2.30.0.tar.gz</span><br><span class="line">cd bedtools2</span><br><span class="line">make</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget https://github.com/samtools/samtools/releases/download/1.13/samtools-1.13.tar.bz2</span><br><span class="line">tar -xvjf samtools-1.13.tar.bz2 </span><br><span class="line">cd samtools-1.13</span><br><span class="line">./configure</span><br><span class="line">make</span><br></pre></td></tr></table></figure>
<p>安装好之后  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">samtools dict chr.fasta &gt; out</span><br><span class="line">cat out | sed &#x27;s/:/\t/g&#x27;|cut -f 3,5 &gt;genome.chr.ln</span><br><span class="line">```  </span><br><span class="line"></span><br><span class="line">- cat out：读取前一步生成的字典文件  </span><br><span class="line">- sed &#x27;s/:/\t/g&#x27;：将行中的冒号替换为制表符（TAB），将信息分隔成多列  </span><br><span class="line">- cut -f 3,5：提取第三列（染色体名称）和第五列（染色体长度）  </span><br><span class="line">- \&gt; genome.chr.ln：&#x27;将结果输出到genome.chr.ln文件中  </span><br><span class="line"></span><br><span class="line">**如果使用circos软件绘图需要使用以下方法**  </span><br></pre></td></tr></table></figure>
<p>genome&#x3D;tair10.genome.fa<br>faidx ${genome} -i chromsizes &gt; tair10.genome.size<br>awk -vFS&#x3D;”\t” -vOFS&#x3D;”\t” ‘{print “chr”,”-“,$1,”C”NR,”0”,$2,”chr”NR}’ tair10.genome.size &gt; circos.chromosome.txt<br>cat circos.chromosome.txt</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">**2）基因密度文件**</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>grep -w “gene” Msin.chr.gff |awk ‘{print $1”\t”$4”\t”$5}’|uniq &gt; gene.pos</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">- grep：用于在文件中查找匹配特定模式的行  </span><br><span class="line">- -w：表示精确匹配整个单词gene，确保只匹配gene（而不是匹配包含gene作为部分的单词，如gene_name等）  </span><br><span class="line">- awk：是一个文本处理工具，常用于提取文件的某些列  </span><br><span class="line">- &#x27;&#123;print $1&quot;\t&quot;$4&quot;\t&quot;$5&#125;&#x27;：这是awk的输出格式，表示提取并输出第1列（染色体）、第4列（基因起始位置）和第5列（基因结束位置），并以制表符 (\t) 分隔  </span><br><span class="line">- uniq：用于删除相邻的重复行。如果文件中的某些基因注释是相同的并且相邻出现，uniq可以确保只保留一个实例  </span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>bedtools makewindows -g genome.chr.ln -w 100000 &gt; 100K.genome.3col</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">- bedtools makewindows：生成固定大小的滑窗（窗口）  </span><br><span class="line">- -g genome.chr.ln：指定包含染色体名称和长度的文件，即前一步生成的genome.chr.ln文件  </span><br><span class="line">- -w 100000：指定滑窗的大小为100,000个碱基对（100kb）  </span><br><span class="line">- \&gt; 100K.genome.3col：将生成的滑窗信息输出到100K.genome.3col文件中  </span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>bedtools intersect -a 100K.genome.3col -b gene.pos -c &gt; gene.out</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">- bedtools intersect：比较两个基因组区间文件，并计算它们之间的重叠</span><br><span class="line">- -a 100K.genome.3col：指定第一个输入文件，这里是包含100kb滑窗的文件（例如100K.genome.3col），这个文件表示基因组上的100kb窗口</span><br><span class="line">- -b gene.pos：指定第二个输入文件，这里是包含基因或其他元素（如LTR、GYPSY、COPIA等）位置的文件（gene.pos）</span><br><span class="line">- -c：计算每个滑窗中与gene.pos中的基因或元素重叠的数量，并将结果添加到输出文件中作为新列</span><br><span class="line">- \&gt; gene.out：将结果输出到文件gene.out中</span><br><span class="line"></span><br><span class="line">如果需要  </span><br></pre></td></tr></table></figure>
<p>sed -i ‘&#x2F;^ptg&#x2F;d’ gene.out</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">-删除序列名中有‘ptg’的序列。这是因为有一些基因组注释文件中包含除染色体之外的基因注释，如我的染色体注释文件中就有ptg开头的序列，因此需要将不相关的统计结果删除</span><br><span class="line"></span><br><span class="line">**3）COPIA密度  </span><br><span class="line">4）GYPSY密度**  </span><br></pre></td></tr></table></figure>
<p>awk ‘{print $1”\t”$4”\t”$5}’  ..&#x2F;JNYL&#x2F;all_chr.gff  |grep “Copia” &gt;copia.bed<br>bedtools intersect -a 100K.genome.3col -b COPIA.pos -c &gt; COPIA.out</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">如果需要  </span><br></pre></td></tr></table></figure>
<p>sed -i ‘&#x2F;^ptg&#x2F;d’ COPIA.out</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">**5）TE密度文件**</span><br><span class="line"></span><br><span class="line">TE密度文件跟重复序列注释文件关系十分密切，如果重复序列注释文件类似以下形式（Repeat Masker注释）:</span><br><span class="line"></span><br><span class="line">    Chr5	RepeatMasker	Transposon	5635057	5638270	25165	+	.	ID=genome_TE0000001;Target=LTR_FINDER_ID=ctg54_3_ltr; 1895 5289;Class=LINE/RTE-BovB;PercDiv=2.7;PercDel=0.2;PercIns=0.3;</span><br><span class="line">    Chr5	RepeatMasker	Transposon	5639304	5639382	422	-	.	ID=genome_TE0000002;Target=LTR_FINDER_ID=ctg54_3_ltr; 3093 3171;Class=LINE/RTE-BovB;PercDiv=16.5;PercDel=0.0;PercIns=0.0;</span><br><span class="line"></span><br><span class="line">就与上面类似，或者可以直接使用注释文件，因为基本全是TE  </span><br></pre></td></tr></table></figure>
<p>grep “Transposon” ..&#x2F;JNYL&#x2F;all_without_trf_chr.gff &gt; TE.gff</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">如果不是，就检索关键词，将Transposon换为TE，或者Transposon Elements等，如果实在弄不清楚，就不要这一个文件，或者尝试使用EDTA，TE Density重新注释TE，构建TE库，但这与本篇教程目的相背离，因此后面再出详细教程</span><br><span class="line"></span><br><span class="line">**6）GC含量文件**</span><br><span class="line"></span><br><span class="line">TBtools-&gt;sequence toolkits-&gt;fasta tools-&gt;fasta Windows stat  </span><br><span class="line">输入文件为chr.fasta，窗口大小设置为10000，相邻窗口之间重叠的碱基数目设置1000，这两参数一般为默认  </span><br><span class="line">指定输出文件路径，运行  </span><br><span class="line">获得的文件为.GCratio文件</span><br><span class="line"></span><br><span class="line">**7）共线性文件**</span><br><span class="line"></span><br><span class="line">共线性文件可以使用服务器构建，也可以使用TBtools构建，在此介绍更为简单的TBtools  </span><br><span class="line">TBtools-&gt;other-&gt;plugin-&gt;plugin store-&gt;sequence analyses-&gt;onestepMCScanX-superfast-&gt;install  </span><br><span class="line">下载好之后onestepMCScanX-superfast位置在TBtools-&gt;other-&gt;plugin-&gt;sequence analyses-&gt;onestepMCScanX-superfast  </span><br><span class="line">第一个和第三个输入文件都为chr.fasta，第二第四个文件均为Msin.chr.gff，指定输出文件路径，运行</span><br><span class="line">最终.geneLinkedRegion.tab.xls文件为共线性文件  </span><br><span class="line">如果要添加颜色，打开共线性文件，在最后一列填入相应的颜色，颜色为RGB格式，可在网站寻找：</span><br><span class="line">https://www.rapidtables.org/zh-CN/web/color/html-color-codes.html#:~:text=%E9%A2%9C%E8%89%B2%20HTML%20/</span><br><span class="line"></span><br><span class="line">例如：</span><br><span class="line"></span><br><span class="line">    Chr1	10556167	10563729	Chr2	72596456	72608457</span><br><span class="line">    Chr1	8479089	8487742	Chr3	8238545	8249992	&quot;255,255,0&quot;</span><br><span class="line">    Chr1	8488084	8491727	Chr3	8256021	8262718	&quot;255,255,0&quot;</span><br><span class="line"></span><br><span class="line">这样可显示chr1和chr3的共线性颜色，chr1和chr2为默认的浅灰色</span><br><span class="line"></span><br><span class="line">使用服务器可使用jcvi建立自比对，此步看之前的教程  </span><br><span class="line">从共线性的基因对，提取整合成circos的bed格式  </span><br></pre></td></tr></table></figure>
<p>cat TNML.TNML.anchors |grep -v #| while read line;<br>do<br>        gene_array&#x3D;($line)<br>        gene1_pos&#x3D;<code>fgrep $&#123;gene_array[0]&#125; TNML.bed|cut -f 1-3</code><br>        gene2_pos&#x3D;<code>fgrep $&#123;gene_array[1]&#125; TNML.bed|cut -f 1-3</code><br>        bed&#x3D;${gene1_pos}” “${gene2_pos}<br>        echo $bed &gt;&gt;TNML.coline_num.txt<br>done</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">## 2.绘制circos图</span><br><span class="line">TBtools-&gt;grapics-&gt;advance circos  </span><br><span class="line">第一个和第三个文件分别输入染色体长度文件和共线性文件，第二个文件不管，运行  </span><br><span class="line">如果想加入其他文件，可通过show control dialog中的add添加，可修改TrackType，有条形线型等，可自己尝试，startpos和endpos中pos指position，可调整位置，不过多赘述  </span><br><span class="line">比较常见要调整的为chr bar和chr label，前者调整染色体长度条形的位置，一般设置在最外层来避免看不清染色体长度，后者调整染色体名字位置，也是同理，其余参数均不需要太更改，因此不过多介绍，可以自己尝试，一般使用AI对输出的PDF文件更改更加方便  </span><br><span class="line"></span><br><span class="line">## 3.使用circos软件绘制（该部分教程参考博客https://www.cnblogs.com/bgicollege/p/15543482.html）</span><br><span class="line">由于Circos是基于perl语言开发的，所以perl语言编译器的安装是必须的（建议Perl 5.8.x及以上版本），除此之外还需要很多perl依赖模块，可以通过测试查看缺少哪些模块并安装。详细的准备要求可以见http://circos.ca/software/requirements/</span><br><span class="line"></span><br><span class="line">### 3.1下载解压</span><br><span class="line">circos的官网的下载页面为http://circos.ca/software/download/，circos-0.69-6.tgz是目前最新版的安装包，而circos-course-2017.tgz则是配套的教程数据。  </span><br></pre></td></tr></table></figure>
<p>wget <a target="_blank" rel="noopener" href="http://circos.ca/distribution/circos-0.69-6.tgz">http://circos.ca/distribution/circos-0.69-6.tgz</a><br>tar xvfz circos-0.69-6.tgz<br>cd circos-0.69-6</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">### 3.2安装缺失依赖模块</span><br><span class="line">直接运行circos，根据报错再依次安装</span><br></pre></td></tr></table></figure>
<p>.&#x2F;circos -v</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">假设出现报错</span><br><span class="line"></span><br><span class="line">    missing Font::TTF::Font</span><br><span class="line">      error Can&#x27;t locate Font/TTF/Font.pm in @INC (you may need to install the Font::TTF::Font module) (@INC contains: /home/walnut168/mnt/wf/program/circos-0.69-9/bin/lib /home/walnut168/mnt/wf/program/circos-0.69-9/bin/../lib /home/walnut168/mnt/wf/program/circos-0.69-9/bin /home/walnut168/perl5/lib/perl5/x86_64-linux-thread-multi /home/walnut168/perl5/lib/perl5 /home/walnut168/miniconda3/envs/wf/lib/perl5/5.32/site_perl /home/walnut168/miniconda3/envs/wf/lib/perl5/site_perl /home/walnut168/miniconda3/envs/wf/lib/perl5/5.32/vendor_perl /home/walnut168/miniconda3/envs/wf/lib/perl5/vendor_perl /home/walnut168/miniconda3/envs/wf/lib/perl5/5.32/core_perl /home/walnut168/miniconda3/envs/wf/lib/perl5/core_perl .) at (eval 35) line 1.</span><br><span class="line">安装缺失依赖模块</span><br></pre></td></tr></table></figure>
<p>cpanm Font::TTF::Font</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">其他缺失模块以此类推</span><br><span class="line"></span><br><span class="line">一般来讲，缺少的模块，大部分安装起来都比较顺利，可能有一个除外，就是GD。GD模块是一个图像相关的源码库，它需要很多前置的依赖包和依赖软件，可部分参考（http://starsyi.github.io/2016/05/31/%E9%9D%9Eroot%E7%94%A8%E6%88%B7perl-GD%E6%A8%A1%E5%9D%97%E7%9A%84%E5%AE%89%E8%A3%85/）</span><br><span class="line">由于年代过于久远，因此参考博客有一部分错误，以下将该流程重新整理一遍。    </span><br><span class="line">（1）下载并编译安装libgd</span><br></pre></td></tr></table></figure>
<p>wget <a target="_blank" rel="noopener" href="https://github.com/libgd/libgd/releases/download/gd-2.3.3/libgd-2.3.3.tar.gz">https://github.com/libgd/libgd/releases/download/gd-2.3.3/libgd-2.3.3.tar.gz</a><br>tar -xzf libgd-2.3.3.tar.gz<br>cd libgd-2.3.3<br>.&#x2F;configure –prefix&#x3D;$HOME&#x2F;local<br>make<br>make install</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">编译安装好之后还需要将将 libgd 的库路径添加到环境变量中</span><br></pre></td></tr></table></figure>
<p>export LD_LIBRARY_PATH&#x3D;$HOME&#x2F;local&#x2F;lib:$LD_LIBRARY_PATH<br>export CPATH&#x3D;$HOME&#x2F;local&#x2F;include:$CPATH<br>export LIBRARY_PATH&#x3D;$HOME&#x2F;local&#x2F;lib:$LIBRARY_PATH</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">（2）下载并安装GD</span><br></pre></td></tr></table></figure>
<p>wget <a target="_blank" rel="noopener" href="https://cpan.metacpan.org/authors/id/R/RU/RURBAN/GD-2.77.tar.gz">https://cpan.metacpan.org/authors/id/R/RU/RURBAN/GD-2.77.tar.gz</a><br>tar -xzf GD-2.77.tar.gz<br>cd GD-2.77<br>perl Makefile.PL –lib_gd_path&#x3D;$HOME&#x2F;local<br>make<br>make test<br>make install PREFIX&#x3D;~&#x2F;perl5</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">安装好之后设置Perl模块路径</span><br></pre></td></tr></table></figure>
<p>export PERL5LIB&#x3D;~&#x2F;perl5&#x2F;lib&#x2F;perl5:$PERL5LIB</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">以上所有操作完成之后运行以下命令验证GD是否安装成功</span><br></pre></td></tr></table></figure>
<p>perl -e ‘use GD;’</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">没有报错说明安装成功，进入circos/bin目录运行以下操作检验circos是否可以正常使用</span><br></pre></td></tr></table></figure>
<p>.&#x2F;circos -v</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br></pre></td><td class="code"><pre><span class="line">没有报错说明安装成功</span><br><span class="line"></span><br><span class="line">### 3.3使用circos</span><br><span class="line">#### 3.3.1准备文件</span><br><span class="line">**（1）.conf文件的介绍**</span><br><span class="line"></span><br><span class="line">圈图中各个环道，是依赖一系列依赖调用的配置文件描述出来的，从底层的颜色等，到更高层，以及最终的元配置，构成了一个堆叠起来的配置文件金字塔。</span><br><span class="line">Circos的配置文件后缀为.conf，内容是标记性文本格式的，类似于HTML或者XML，每个标签对代表一个配置块，起始标记为\&lt;TAG&gt;，终止标记为\&lt;/TAG&gt;。配置块中的基本语法就是:</span><br><span class="line"></span><br><span class="line">     variable = value</span><br><span class="line"></span><br><span class="line">举一个例子，一个简单的配置块：</span><br><span class="line"></span><br><span class="line">    &lt;plots&gt;</span><br><span class="line">    type = line</span><br><span class="line">    thickness = 2</span><br><span class="line">    ……</span><br><span class="line">    &lt;plot&gt;</span><br><span class="line"></span><br><span class="line">每个标签对可以使用多次，或者相互嵌套。因此就像编程中的函数或者循环结构一样，有全局变量和局部变量的区分，高级的或者外层的变量对于层内的所有实例都是有效的，而层内的定义则是优先级最高的。例如，下面配置中的  fill_color = blue   对于对  \&lt;plots&gt;\&lt;/plots&gt;   中所有实例都有效，除非实例内部自己用局部变量进行定义，如内部嵌套的第一个    \&lt;plot&gt;\&lt;/plot&gt; 对中的  fill_color = red    ：</span><br><span class="line"></span><br><span class="line">    &lt;plots&gt;</span><br><span class="line">    fill_color = blue</span><br><span class="line">    ……</span><br><span class="line">        &lt;plot&gt;</span><br><span class="line">        fill_color = red</span><br><span class="line">        ……</span><br><span class="line">        &lt;/plot&gt;</span><br><span class="line"></span><br><span class="line">        &lt;plot&gt;</span><br><span class="line">        ……</span><br><span class="line">        &lt;/plot&gt;</span><br><span class="line">    &lt;/plots&gt;</span><br><span class="line"></span><br><span class="line">一些围绕同一个主题的重要配置块，可以另外单独写到一个配置文件中，再由其他的模块导入调用，所以理论上我们可以把所有的配置写在同一个文件中，也可以把每个配置块都分开到不同的文件中写，最后统一调用。</span><br><span class="line"></span><br><span class="line">其中最基本的配置文件都放在$PATH/circos-0.69-4/etc下面，主要是：</span><br><span class="line"></span><br><span class="line">    Housekeeping.conf #基本框架配置，circos必不可少的配置文件，一般直接导入即可。</span><br><span class="line">    fonts.conf        #字体配置文件，其中，代表字体的标签与circos/fonts文件夹下的otf或ttf文件一一对应。</span><br><span class="line">    patterns.conf     #模式配置文件，其中，标签与circos/tites文件夹下的png文件一一对应，然后根据patterns.svg.conf生成相应的svg文件。</span><br><span class="line">    colors.conf       #颜色配置文件，导入了colors.brewer.conf、colors.hsv.conf和colors.ucsc.conf。在circos中，颜色除了绝对路径，还有很多表达方式，如hs1，red，reds-9-seq-1等，是因为这些字符在colors配置文件中已经赋予了特定的绝对路径。</span><br><span class="line">以上的基本配置大多数情况下无需调整，直接调用导入即可。导入配置的基本格式是：</span><br><span class="line"></span><br><span class="line">    &lt;&lt;include etc/*.conf&gt;&gt; #etc/*.conf就是要导入的配置文件的相对或者绝对路径</span><br><span class="line">还有一些经常用到的其他配置，可以根据自己喜好定制：</span><br><span class="line"></span><br><span class="line">ideogram.conf  #染色体展示的配置文件，包括是否展示bands，染色体展示位置、染色体间距离、染色体标签位置等。</span><br><span class="line">ticks.conf     #刻度标签配置文件，可以导入到ideogram.conf中，也可以直接导入到主配置文件中。</span><br><span class="line">image.conf     #对image大小、背景颜色、输出目录、输出文件形式以及染色体在圆周上的起始位置等进行设置。</span><br><span class="line">最后，这些配置文件都会导入到circos.conf，基本内容是：</span><br><span class="line"></span><br><span class="line">    karyotype = data/karyotype/karyotype.mouse.mm9.txt #核型文件</span><br><span class="line">    chromosomes_units = 1000000 #就是其他的配置文件中距离单位u的大小，以bp来衡量，一般大的基因组倾向于把一个单位的长度设置得大一些</span><br><span class="line">    chromosomes_display_default = yes</span><br><span class="line">    &lt;&lt;include mydata/ideogram.conf&gt;&gt;</span><br><span class="line">    &lt;&lt;include mydata/ticks.conf&gt;&gt;</span><br><span class="line">    ……</span><br><span class="line">    &lt;image&gt;</span><br><span class="line">    &lt;&lt;include etc/image.conf&gt;&gt;</span><br><span class="line">    &lt;/image&gt;</span><br><span class="line">    &lt;&lt;include etc/colors_fonts_patterns.conf&gt;&gt;</span><br><span class="line">    &lt;&lt;include etc/housekeeping.conf&gt;&gt;</span><br><span class="line">配置完所有的配置文件后，最后将这个元配置文件输入Circos就可以产生结果：</span><br></pre></td></tr></table></figure>
<p>circos -conf circos.conf</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">**（2）基因组作图文件**</span><br><span class="line"></span><br><span class="line">1.计算染色体长度  </span><br><span class="line">安装samtools  </span><br></pre></td></tr></table></figure>
<p>wget <a target="_blank" rel="noopener" href="https://github.com/samtools/samtools/releases/download/1.13/samtools-1.13.tar.bz2">https://github.com/samtools/samtools/releases/download/1.13/samtools-1.13.tar.bz2</a><br>tar -xvjf samtools-1.13.tar.bz2<br>cd samtools-1.13<br>.&#x2F;configure<br>make</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">使用samtools中的插件faidx  </span><br></pre></td></tr></table></figure>
<p>faidx ..&#x2F;zjll&#x2F;ZJ.chr.fa -i chromsizes &gt; zj.size<br>awk -vFS&#x3D;”\t” -vOFS&#x3D;”\t” ‘{print “chr”,”-“,$1,”C”NR,”0”,$2,”chr”NR}’ zj.size &gt; zj.circos.chromosome.txt</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">也可以使用seqtk  </span><br></pre></td></tr></table></figure>
<p>seqtk comp ..&#x2F;zjll&#x2F;ZJ.chr.fa | grep “^chr” | awk ‘{print $1”\t”$2}’ &gt; zj.genome.len<br>awk ‘{print “chr\t-\t”$1”\t”$1”\t0\t”$2”\tchr”NR}’ zj.genome.len &gt; zj_karyotype.txt<br>#此处zj_karyotype.txt等同于上一步中的zj.circos.chromosome.txt，zj.size等同于zj.genome.len，两者均可相互替换，后不赘述</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">zj.circos.chromosome.txt结果如下：</span><br><span class="line"></span><br><span class="line">    #染色体 -   染色体编号(ID) 图显示名称(Label)     起始   结束    颜色</span><br><span class="line">    chr     -       chr1    C1      0       33953567        chr1</span><br><span class="line">    chr     -       chr2    C2      0       36348176        chr2</span><br><span class="line">    chr     -       chr3    C3      0       32768115        chr3</span><br><span class="line">    chr     -       chr4    C4      0       22859102        chr4</span><br><span class="line">    chr     -       chr5    C5      0       19899253        chr5</span><br><span class="line">    chr     -       chr6    C6      0       28512749        chr6</span><br><span class="line">    chr     -       chr7    C7      0       28542370        chr7</span><br><span class="line">    chr     -       chr8    C8      0       36419750        chr8</span><br><span class="line">    chr     -       chr9    C9      0       33812785        chr9</span><br><span class="line">    chr     -       chr10   C10     0       23968779        chr10</span><br><span class="line">    chr     -       chr11   C11     0       29606729        chr11</span><br><span class="line">    chr     -       chr12   C12     0       38801854        chr12</span><br><span class="line">    chr     -       chr13   C13     0       49412298        chr13</span><br><span class="line">    chr     -       chr14   C14     0       31566911        chr14</span><br><span class="line">    chr     -       chr15   C15     0       37966322        chr15</span><br><span class="line">    chr     -       chr16   C16     0       43438598        chr16</span><br><span class="line"></span><br><span class="line">2.GC含量  </span><br><span class="line">生成窗口文件</span><br></pre></td></tr></table></figure>
<p>bedtools makewindows -w 50000 -g zj.genome.len &gt; zj.window.bed</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">计算每个窗口平均GC含量  </span><br></pre></td></tr></table></figure>
<p>seqtk subseq ..&#x2F;zjll&#x2F;ZJ.chr.fa zj.window.bed &gt; zj.window.fa<br>seqtk comp zj.window.fa |awk ‘{print $1 “\t” ($4+$5)&#x2F;($3+$4+$5+$6) } ‘ | awk -F “:|-“ ‘{print $1”\t”$2”\t”$3”\t”$4}’&gt; zj_gc.txt</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">3.基因条数  </span><br></pre></td></tr></table></figure>
<p>grep -w “gene” LV.chr.gff |awk ‘{print $1”\t”$4”\t”$5}’|uniq &gt; gene.bed</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">```</span><br><span class="line">bedtools coverage -a genome.window.bed-b gene.bed -c &gt; gene.out</span><br></pre></td></tr></table></figure>
<p>4.重复序列含量</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">bedtools coverage -a genome.window.bed -b repeat.gff | awk &#x27;&#123;print $1 &quot;\t&quot; $2 &quot;\t&quot; $3 &quot;\t&quot; $7&#125;&#x27; &gt; zj_repeat.txt</span><br><span class="line">#copia、gypsy</span><br><span class="line">grep &quot;Copia&quot;   ../JNYL/all_chr.gff |awk &#x27;&#123;print $1&quot;\t&quot;$4&quot;\t&quot;$5&#125;&#x27;  &gt;copia.bed</span><br><span class="line">bedtools coverage -a genome.window.bed-b copia.bed -c &gt; copia.out</span><br></pre></td></tr></table></figure>
<p>5.共线性link文件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cat ZJ.LV.anchors | grep -v \# | while read line;</span><br><span class="line">do</span><br><span class="line">    gene_array=($line)</span><br><span class="line">    gene1_pos=$(awk -v gene=&quot;$&#123;gene_array[0]&#125;&quot; &#x27;$4 == gene &#123;print $1, $2, $3&#125;&#x27; zj.bed)</span><br><span class="line">    gene2_pos=$(awk -v gene=&quot;$&#123;gene_array[1]&#125;&quot; &#x27;$4 == gene &#123;print $1, $2, $3&#125;&#x27; LV.bed)</span><br><span class="line">    bed=$&#123;gene1_pos&#125;&quot; &quot;$&#123;gene2_pos&#125;</span><br><span class="line">    echo $bed &gt;&gt; ZL.LINKS.txt</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">cat LV.LV.anchors |grep -v \#| while read line;</span><br><span class="line">do</span><br><span class="line">        gene_array=($line)</span><br><span class="line">        gene1_pos=`fgrep $&#123;gene_array[0]&#125; LV.bed|cut -f 1-3`</span><br><span class="line">        gene2_pos=`fgrep $&#123;gene_array[1]&#125; LV.bed|cut -f 1-3`</span><br><span class="line">        bed=$&#123;gene1_pos&#125;&quot; &quot;$&#123;gene2_pos&#125;</span><br><span class="line">        echo $bed &gt;&gt;LV.LINKS.txt</span><br><span class="line">done</span><br><span class="line"># 使用 awk 重组数据（假设原始数据每行包含6的倍数列）</span><br><span class="line">awk &#x27;&#123;</span><br><span class="line">    for(i=1;i&lt;=NF;i+=6) &#123;</span><br><span class="line">        if($(i+5)) print $i,$(i+1),$(i+2),$(i+3),$(i+4),$(i+5)</span><br><span class="line">    &#125;</span><br><span class="line">&#125;&#x27; LV.LINKS.txt &gt; LV.LINKS.circos.txt</span><br></pre></td></tr></table></figure>

<p><strong>（3）核型配置文件</strong></p>
<pre><code>Circos的配置中有四种单位，p, r, u, b
p表示像素，1p表示1像素
r表示相对大小，0.95r表示95% ring 大小。
u表示相对chromosomes_unit的长度，如果chromosomes_unit = 1000，则1u就是千分之一的染色体长度。
b表示碱基，如果染色体长1M，那么1b就是百万分之一的长度。
</code></pre>
<p>染色体图形配置文件，ideogram.conf文件基本配置如下：</p>
<pre><code>&lt;ideogram&gt;
    show = yes
    &lt;spacing&gt;
        default = 0.005r    #染色体间的空隙
    &lt;/spacing&gt;
    thickness      = 25p    #ideograms的厚度，可以使用r（比例关系）或p（像素）作为单位
    fill           = yes    #设定ideograms是否填充颜色。填充的颜色取决于 karyotype 指定的文件的最后一列。
    radius         = 0.8r   #设定ideograms的位置
    show_label     = yes    #设定是否显示label
    label_font     = default#设定label的字体
    label_radius   = dims(ideogram,radius_outer) + 80p #设定label的位置
    label_size     = 40p    #设定label的字体大小
    label_parallel = yes    #设定label的字体方向，yes是易于浏览的方向。
&lt;/ideogram&gt;
</code></pre>
<p>刻度文件，ticks.conf，基本配置如下：</p>
<pre><code>&lt;ticks&gt;
    tick_label_font  = light
    radius           = dims(ideogram,radius_outer) + 10p  # 设定 ticks 的位置
    label_offset     = 5p  # 设置 ticks&#39; label 离 ticks 的距离
    label_size       = 20p  # 设置 ticks&#39; label 的字体大小
    multiplier       = 0.001  # 设定 ticks&#39; label 的值的计算。将该刻度对应位置的值 * multiplier 得到能展示到圈图上的 label 值。
    color            = black  # 设定 ticks 的颜色
    thickness        = 1p  # 设定 ticks 的厚度

    &lt;tick&gt;
        spacing        = 50u  # 设置每个刻度代表的长度。
        size           = 12p  # 设置 tick 的长度
        show_label     = yes  # 是否显示 tick labels
    &lt;/tick&gt;

    &lt;tick&gt;
        label_separation = 1p
        spacing          = 5u
        size             = 7p
        show_label       = no
    &lt;/tick&gt;
&lt;/ticks&gt;
</code></pre>
<p>主配置文件，circos.conf，主要配置如下：</p>
<pre><code>karyotype = karyotype.txt #指定染色体文件
chromosomes_units = 1000000 #设置长度单位，表示为1M长度的序列代表为1u
chromosomes_display_default = yes #默认是将所有的染色体都展示出来
## 载入ideogram配置和刻度线配置
show_ticks = yes # 显示刻度
show_tick_labels = yes # 显示刻度label
## plots block 绘制折线图、散点图、直方图、热图和文本显示
&lt;plots&gt;
## 绘制直方图
&lt;plot&gt;
type = histogram # histogram, line,scatter, heatmap,text
file = data/gene_density.txt #文件路径
# 设置直方图的位置，r1 &gt; r0
r1 = 0.98r
r0 = 0.9r
orientation = out #直方图方向，in为向内
fill_color = blue #直方图的填充颜色
thickness = 2p #直方图轮廓厚度为 1p。
extend_bin = no #若bins在坐标上不相连，最好设置不要将其连接到一起。
# 设定直方图的背景颜色
&lt;backgrounds&gt;
&lt;background&gt;
color = vlgrey
&lt;/background&gt;
&lt;/backgrounds&gt;
&lt;/plot&gt;
##绘制热图
&lt;plot&gt;
type = heatmap # 指定绘图类型为heatmap
file = data/gene_density.txt # 设定数据文件路径
# 设定图形所处位置
r1 = 0.88r
r0 = 0.8r
color = green,yellow,red # 设定热图的颜色
scale_log_base = 1 # 设定 scale_log_base 参数
&lt;/plot&gt;
##绘制折线图
&lt;plot&gt;
type = line # 指定绘图类型为line
file = data/gene_density.txt # 设定数据文件路径
# 设定图形所处位置
r1 = 0.78r
r0 = 0.7r
color = green # 设定颜色
&lt;backgrounds&gt;
&lt;background&gt;
y1 = 0.5r
y0 = 0r
color = vvlgrey
&lt;/background&gt;
&lt;background&gt;
y1 = 1r
y0 = 0.5r
color = vvlred
&lt;/background&gt;
&lt;/backgrounds&gt;
&lt;/plot&gt;
##绘制散点图
&lt;plot&gt;
type = scatter # 指定绘图类型为scatter
file = data/gene_density.txt # 设定数据文件路径
# 设定图形所处位置
r1 = 0.68r
r0 = 0.6r
glyph = circle # circle, rectangle, or triangle
glyph_size = 4p # 点大小
color = green # 设定颜色
&lt;backgrounds&gt;
&lt;background&gt;
color = vvlgrey
&lt;/background&gt;
&lt;/backgrounds&gt;
&lt;/plot&gt;
##显示文本
&lt;plot&gt;
type = text #设置绘图类型为文本
file = data/text.gene.txt # 数据文件路径
color = black #文字颜色
# 显示在图形中的位置
r1 = 0.58r
r0 = 0.45r
label_font = default # 标签的字体
label_size = 20p # 标签大小
rpadding = 1p # 文字边缘的大小
# 设置是否需要在 label 前加一条线，用来指出 lable 的位置。
show_links = yes
link_dims = 0p,2p,5p,2p,2p
link_thickness = 2p
link_color = red
&lt;/plot&gt;
&lt;/plots&gt;
##links画连接线
&lt;links&gt;
&lt;link&gt;
file = data/links.genePair.txt #link文件
color = blue #设置颜色
radius = 0.6r #设置 link 曲线的半径
bezier_radius = 0r #设置贝塞尔曲线半径，该值设大后曲线扁平。
thickness = 2 #设置 link 曲线的厚度
## ribbon = yes #将线条连接改成带状连接
######rules指定线条的颜色，
&lt;rules&gt;
&lt;rule&gt;
condition = var(chr1) eq &quot;Chr1&quot; #共线性文件中，左边是Chr1的，指定颜色。var(chr1)这个不用修改，指的是file里左侧第一个，后面的“Chr1”根据实际的Chr来修改。
color = rdylgn-11-div-1
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr2&quot;
color = rdylgn-11-div-2
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr3&quot;
color = rdylgn-11-div-3
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr4&quot;
color = rdylgn-11-div-4
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr5&quot;
color = rdylgn-11-div-5
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr6&quot;
color = rdylgn-11-div-6
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr7&quot;
color = rdylgn-11-div-7
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr8&quot;
color = rdylgn-11-div-8
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr9&quot;
color = rdylgn-11-div-9
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr10&quot;
color = rdylgn-11-div-10
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr11&quot;
color = rdylgn-11-div-11
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr12&quot;
color = rdylbu-9-div-1
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr13&quot;
color = rdylbu-9-div-2
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr14&quot;
color = rdylbu-9-div-3
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr15&quot;
color = rdylbu-9-div-4
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr16&quot;
color = rdylbu-9-div-5
&lt;/rule&gt;

&lt;rule&gt;
condition = var(chr1) eq &quot;Chr17&quot;
color = rdylbu-9-div-6
&lt;/rule&gt;
&lt;/link&gt;
&lt;/links&gt;
&lt;ideogram&gt;
&lt;spacing&gt;
default = 0.005r

#设置染色体之间空出个距离，用来标图例a,b,c,d,e,f
&lt;pairwise Chr17;Chr1&gt;  #染色体名称是karyotype.Spo.txt中第3列的名称，根据你实际的物种第一条和最后一条来修改此处
spacing = 5r #设置第一个和最后一个染色体中间空出5*default（0.005r）的距离
&lt;/pairwise&gt;

&lt;/spacing&gt;

radius           = 0.85r #指定画图的半径（此半径和plots里的半径不一致，这个是总的0.85r，决定了整个圈图的最大的半径,plots里的0.99r半径是以这个0.85r作为100%来分配99%）
thickness        = 10p #坐标轴的染色体的厚度
fill             = yes
show_bands      = yes #设置染色体条带填充，如果不想用条带填充，只用颜色，就设置为no
fill_bands      = yes
stroke_color     = dgrey
stroke_thickness = 2p
show_label     = yes #展示label
label_font     = default # 字体
label_radius   = dims(ideogram,radius) + 0.075r #位置
label_size     = 12 # 字体大小
label_parallel = yes # 是否平行
label_format   = eval(sprintf(&quot;%s&quot;,var(chr))) # 格式
&lt;/ideogram&gt;

##以下部分通常不做修改
#设置图片参数
&lt;image&gt;
#覆盖原来的图片半径参数
#radius* = 500
&lt;/image&gt;
#设置颜色，字体，填充模式的配置信息：
#系统与debug参数：
</code></pre>
<h4 id="3-3-3运行"><a href="#3-3-3运行" class="headerlink" title="3.3.3运行"></a>3.3.3运行</h4><p>完成以上配置文件后运行命令</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">circos -conf circos.conf</span><br></pre></td></tr></table></figure>
<p>即可<br>如果是两个物种之间的比较的圈图，直接将对应的文件合并成一个文件即可，需要注意的是，两个物种的染色体名称需要不同</p>
<p><em><strong>&lt;后话&gt;</strong></em><br>circos图可展示的远不止以上这些，还可以展示染色体可及性，表达水平，只要最后文件与上面所展示的文件格式类似就可以展示，不得不感慨，这真的是技术与美学的一场碰撞。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2025/04/17/hello-world/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="Plank">
      <meta itemprop="description" content="白马？定叫他有来无回！">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="WeFun blog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2025/04/17/hello-world/" class="post-title-link" itemprop="url">Hello World</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2025-04-17 20:36:09" itemprop="dateCreated datePublished" datetime="2025-04-17T20:36:09+08:00">2025-04-17</time>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>Welcome to <a target="_blank" rel="noopener" href="https://hexo.io/">Hexo</a>! This is your very first post. Check <a target="_blank" rel="noopener" href="https://hexo.io/docs/">documentation</a> for more info. If you get any problems when using Hexo, you can find the answer in <a target="_blank" rel="noopener" href="https://hexo.io/docs/troubleshooting.html">troubleshooting</a> or you can ask me on <a target="_blank" rel="noopener" href="https://github.com/hexojs/hexo/issues">GitHub</a>.</p>
<h2 id="Quick-Start"><a href="#Quick-Start" class="headerlink" title="Quick Start"></a>Quick Start</h2><h3 id="Create-a-new-post"><a href="#Create-a-new-post" class="headerlink" title="Create a new post"></a>Create a new post</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ hexo new <span class="string">&quot;My New Post&quot;</span></span><br></pre></td></tr></table></figure>

<p>More info: <a target="_blank" rel="noopener" href="https://hexo.io/docs/writing.html">Writing</a></p>
<h3 id="Run-server"><a href="#Run-server" class="headerlink" title="Run server"></a>Run server</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ hexo server</span><br></pre></td></tr></table></figure>

<p>More info: <a target="_blank" rel="noopener" href="https://hexo.io/docs/server.html">Server</a></p>
<h3 id="Generate-static-files"><a href="#Generate-static-files" class="headerlink" title="Generate static files"></a>Generate static files</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ hexo generate</span><br></pre></td></tr></table></figure>

<p>More info: <a target="_blank" rel="noopener" href="https://hexo.io/docs/generating.html">Generating</a></p>
<h3 id="Deploy-to-remote-sites"><a href="#Deploy-to-remote-sites" class="headerlink" title="Deploy to remote sites"></a>Deploy to remote sites</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">$ hexo deploy</span><br></pre></td></tr></table></figure>

<p>More info: <a target="_blank" rel="noopener" href="https://hexo.io/docs/one-command-deployment.html">Deployment</a></p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">Plank</p>
  <div class="site-description" itemprop="description">白马？定叫他有来无回！</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">2</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
  </nav>
</div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Plank</span>
</div>
  <div class="powered-by">由 <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a> 强力驱动
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

</body>
</html>
