<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 7.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/1.svg">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/1.svg">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="查找共有独有基因为什么要查找共有和独有基因？共有基因一般是同源基因，独有基因是指该物种因为环境而被迫演化出的的基因，这意味这在对应的环境下，必定有一些适应该环境的特殊基因，这可以帮助我们弄清楚物种的之间的演化过程，也可以在育种方面提供原材料。共有基因往往对应的是一些基因家族，在漫长的演化过程中，基因家族中有一些基因减少，或者增多，这其实就对应着基因家族的扩张收缩，这也是研究物种演化的关键数据之一。">
<meta property="og:type" content="article">
<meta property="og:title" content="WeFun blog">
<meta property="og:url" content="http://example.com/2025/05/19/%E6%9F%A5%E6%89%BE%E5%85%B1%E6%9C%89%E5%92%8C%E7%8B%AC%E6%9C%89%E5%9F%BA%E5%9B%A0/index.html">
<meta property="og:site_name" content="WeFun blog">
<meta property="og:description" content="查找共有独有基因为什么要查找共有和独有基因？共有基因一般是同源基因，独有基因是指该物种因为环境而被迫演化出的的基因，这意味这在对应的环境下，必定有一些适应该环境的特殊基因，这可以帮助我们弄清楚物种的之间的演化过程，也可以在育种方面提供原材料。共有基因往往对应的是一些基因家族，在漫长的演化过程中，基因家族中有一些基因减少，或者增多，这其实就对应着基因家族的扩张收缩，这也是研究物种演化的关键数据之一。">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2025-05-19T07:53:23.598Z">
<meta property="article:modified_time" content="2025-05-19T07:49:43.689Z">
<meta property="article:author" content="Plank">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/2025/05/19/%E6%9F%A5%E6%89%BE%E5%85%B1%E6%9C%89%E5%92%8C%E7%8B%AC%E6%9C%89%E5%9F%BA%E5%9B%A0/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'zh-CN'
  };
</script>

  <title> | WeFun blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">WeFun blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>归档</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2025/05/19/%E6%9F%A5%E6%89%BE%E5%85%B1%E6%9C%89%E5%92%8C%E7%8B%AC%E6%9C%89%E5%9F%BA%E5%9B%A0/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Plank">
      <meta itemprop="description" content="白马？定叫他有来无回！">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="WeFun blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>
              

              <time title="创建时间：2025-05-19 15:53:23 / 修改时间：15:49:43" itemprop="dateCreated datePublished" datetime="2025-05-19T15:53:23+08:00">2025-05-19</time>
            </span>

          <br>
            <span class="post-meta-item" title="本文字数">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">本文字数：</span>
              <span>6.6k</span>
            </span>
            <span class="post-meta-item" title="阅读时长">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">阅读时长 &asymp;</span>
              <span>6 分钟</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h1 id="查找共有独有基因"><a href="#查找共有独有基因" class="headerlink" title="查找共有独有基因"></a>查找共有独有基因</h1><p><strong>为什么要查找共有和独有基因？</strong><br>共有基因一般是同源基因，独有基因是指该物种因为环境而被迫演化出的的基因，这意味这在对应的环境下，必定有一些适应该环境的特殊基因，这可以帮助我们弄清楚物种的之间的演化过程，也可以在育种方面提供原材料。<br>共有基因往往对应的是一些基因家族，在漫长的演化过程中，基因家族中有一些基因减少，或者增多，这其实就对应着基因家族的扩张收缩，这也是研究物种演化的关键数据之一。<br>查找共有和独有基因其实就是做基因家族的聚类，分析基因家族的扩张收缩，常有的软件有orthofinder和orthomcl，orthofinder的集成度更好，还可以自动完成建树因此此处我们学习orthofinder的使用。  </p>
<h2 id="1-安装"><a href="#1-安装" class="headerlink" title="1.安装"></a>1.安装</h2><p>下载和官方教程地址：<a target="_blank" rel="noopener" href="https://github.com/davidemms/OrthoFinder">https://github.com/davidemms/OrthoFinder</a><br>可以使用conda安装  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">conda install orthofinder -c bioconda</span><br></pre></td></tr></table></figure>
<p>也可以手动安装  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">wget https://github.com/davidemms/OrthoFinder/releases/latest/download/OrthoFinder.tar.gz</span><br><span class="line">tar xzvf OrthoFinder.tar.gz</span><br></pre></td></tr></table></figure>
<h2 id="2-准备数据"><a href="#2-准备数据" class="headerlink" title="2.准备数据"></a>2.准备数据</h2><p>本次所使用的数据是Magnolia_biondii.chr.pep和Msin.chr.pep两物种的蛋白质数据，但这两个数据往往不可以直接获得，就算直接获得的，也往往不是最长转录本，因此我们先要进行一些处理<br>首先需要根据数据提取最长转录本<br><strong>为什么要提取最长转录本？</strong><br>由于可变剪切的存在，通常一个基因可以转录为多个转录本。但是如果将多个转录本同时进行分析，那么分析会因此受到影响。所以，目前的解决办法是，选取一个最具代表性的转录本（最长转录本）来进行分析。<br>最长转录本其实可以从序列文件中获取，也可以从gff文件中获取，获取的方式也有很多，可以直接使用gettranstool，用</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">GetLongestTransFromGencode --file example.fa --outfile longest_trans_gencode.fa</span><br></pre></td></tr></table></figure>
<p>官方的教程：<a target="_blank" rel="noopener" href="https://github.com/junjunlab/GetTransTool%EF%BC%8C%E6%9C%89%E9%9C%80%E8%A6%81%E8%87%AA%E5%B7%B1%E5%8F%AF%E4%BB%A5%E6%9F%A5%E9%98%85">https://github.com/junjunlab/GetTransTool，有需要自己可以查阅</a><br>在此介绍对于小白最友好的方法，即使用之前提过多次的软件，TBtools：  </p>
<pre><code>sequence toolkit-&gt;fasta tools-&gt;fasta get representative  
</code></pre>
<p>note部分不需要管，ID grouping pattern处为正则表达式，这是序列名称的编号规则，一般不需要更改，然后是输入蛋白质序列数据以及输出文件名的设置<br>建议直接使用TBtools最简单，只需要弄清楚ID号的正则表达式即可  </p>
<p>##蛋白质序列的序列名应该尽量简单，否则可能会出现报错</p>
<pre><code>ValueError: too many values to unpack
</code></pre>
<p>这种情况一般在10个以上物种进行聚类的时候出现，出现这种情况时，可以直接使用python脚本</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">#!/usr/bin/env python3</span><br><span class="line">import os</span><br><span class="line">import re</span><br><span class="line">from Bio import SeqIO</span><br><span class="line">import sys</span><br><span class="line"></span><br><span class="line">def simplify_fasta_headers(input_file, output_file):</span><br><span class="line">    &quot;&quot;&quot;简化单个FASTA文件的序列头，保留第一个有效标识符&quot;&quot;&quot;</span><br><span class="line">    with open(output_file, &quot;w&quot;) as out_handle:</span><br><span class="line">        for record in SeqIO.parse(input_file, &quot;fasta&quot;):</span><br><span class="line">            new_id = record.id.split()[0]  # 取第一个空格前的部分</span><br><span class="line">            record.id = new_id</span><br><span class="line">            record.description = new_id</span><br><span class="line">            SeqIO.write(record, out_handle, &quot;fasta&quot;)</span><br><span class="line"></span><br><span class="line">def process_directory(input_dir, output_dir):</span><br><span class="line">    &quot;&quot;&quot;批量处理目录下所有.fasta文件&quot;&quot;&quot;</span><br><span class="line">    if not os.path.exists(output_dir):</span><br><span class="line">        os.makedirs(output_dir)</span><br><span class="line"></span><br><span class="line">    for filename in os.listdir(input_dir):</span><br><span class="line">        if filename.endswith(&quot;.fasta&quot;) or filename.endswith(&quot;.fa&quot;):</span><br><span class="line">            input_path = os.path.join(input_dir, filename)</span><br><span class="line">            output_path = os.path.join(output_dir, filename)</span><br><span class="line">            simplify_fasta_headers(input_path, output_path)</span><br><span class="line">            print(f&quot;处理完成: &#123;filename&#125; -&gt; &#123;output_path&#125;&quot;)</span><br><span class="line"></span><br><span class="line">if __name__ == &quot;__main__&quot;:</span><br><span class="line">    if len(sys.argv) != 3:</span><br><span class="line">        print(f&quot;用法: &#123;sys.argv[0]&#125; &lt;输入文件夹&gt; &lt;输出文件夹&gt;&quot;)</span><br><span class="line">        print(f&quot;示例: &#123;sys.argv[0]&#125; ./input_fasta ./cleaned_fasta&quot;)</span><br><span class="line">        sys.exit(1)</span><br><span class="line"></span><br><span class="line">    input_dir = sys.argv[1]</span><br><span class="line">    output_dir = sys.argv[2]</span><br><span class="line">    process_directory(input_dir, output_dir)</span><br><span class="line">    print(f&quot;\n所有文件处理完成！结果保存在: &#123;output_dir&#125;&quot;)</span><br></pre></td></tr></table></figure>
<p>该脚本是终端脚本，使用方式为</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python simplify_headers.py ./原始文件夹 ./输出文件夹</span><br></pre></td></tr></table></figure>
<p>删除所有非标准氨基酸字符</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">for i in data4/*;do sed -i &#x27;/^&gt;/! s/[^A-Za-z]//g&#x27; &quot;$i&quot;; done</span><br></pre></td></tr></table></figure>
<h2 id="3-家族聚类及结果解读"><a href="#3-家族聚类及结果解读" class="headerlink" title="3.家族聚类及结果解读"></a>3.家族聚类及结果解读</h2><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">orthofinder -f data -S diamond -M msa -T fasttree -t 20 </span><br><span class="line">-f： 数据目录，先将所有蛋白放进文件夹data中</span><br><span class="line">-S： 比对方法，blast, mmseqs, blast_gz, diamond</span><br><span class="line">-M： 基因树推断方法，dendroblast，msa</span><br><span class="line">-T：建树软件，iqtree, raxml-ng, fasttree, raxml</span><br><span class="line">-t： 线程数，根据服务器配置选择，默认为56</span><br></pre></td></tr></table></figure>
<p>聚类结果如下：</p>
<pre><code>Results_XX #结果文件夹名
Citation.txt
Comparative_Genomics_Statistics #统计结果
Gene_Duplication_Events
Gene_Trees # 每个家族基因树
Log.txt
MultipleSequenceAlignments # 多序列比对结果 包含单拷贝基因supergene
Orthogroups # 主要结果目录
Orthogroup_Sequences # 每个家族序列文件
Orthologues
Phylogenetically_Misplaced_Genes
Phylogenetic_Hierarchical_Orthogroups
Putative_Xenologs
Resolved_Gene_Trees
Single_Copy_Orthologue_Sequences # 单拷贝基因家族序列
Species_Tree # 物种树构建结果
WorkingDirectory

Results_XX/Orthogroups#主要结果
Orthogroups.GeneCount.tsv # 基因数目
Orthogroups_SingleCopyOrthologues.txt # 单拷贝基因家族列表
Orthogroups.tsv # 聚类结果
Orthogroups.txt # 聚类结果
Orthogroups_UnassignedGenes.tsv # 未聚类基因
</code></pre>
<h2 id="4-处理聚类结果并绘制韦恩图"><a href="#4-处理聚类结果并绘制韦恩图" class="headerlink" title="4.处理聚类结果并绘制韦恩图"></a>4.处理聚类结果并绘制韦恩图</h2><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">awk &#x27;&#123; if(NR==1)&#123; for(i=2;i&lt;NF;i++ )&#123;printf $i&quot;\t&quot;&#125; &#125;else&#123; \</span><br><span class="line">for(i=2;i&lt;NF;i++)&#123; if($i&gt;0)&#123;printf $1 &#125; ; printf &quot;\t&quot; &#125;&#125;; \</span><br><span class="line">printf &quot;\n&quot;&#125;&#x27; Orthogroups.GeneCount.tsv | \</span><br><span class="line">sed &#x27;s/\t$//&#x27; &gt; Orthogroups.GeneCount.venn</span><br></pre></td></tr></table></figure>
<p>脚本解释：  </p>
<pre><code>NR==1: 当读取到文件的第一行时（通常是列标题行），执行以下操作：
    for(i=2;i&lt;NF;i++): 从第2列开始（跳过第1列），循环处理每一列直到倒数第2列。
    printf $i&quot;\t&quot;: 输出每列的内容，后面跟一个制表符（\t），这些输出结果是列标题（从第2列开始），将用于标记维恩图中不同组的名称。
    else&#123; for(i=2;i&lt;NF;i++)&#123; if($i&gt;0)&#123;printf $1 &#125;; printf &quot;\t&quot; &#125;&#125;
非第一行时
    for(i=2;i&lt;NF;i++): 对每一行，从第2列开始直到倒数第2列进行循环。
    if($i&gt;0)&#123;printf $1 &#125;: 如果当前列的值大于0，则输出该行的第1列内容（通常是一个基因组的名称或ID）。
    printf &quot;\t&quot;: 在输出内容之后，加入一个制表符。
</code></pre>
<p>每处理完一行后，脚本在行尾输出一个换行符（\n），以确保每行的数据换行。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">sed &#x27;s/\t$//&#x27;</span><br><span class="line">    sed命令用于对awk生成的输出进行后处理。s/\t$//是一个替换操作：</span><br><span class="line">        \t$表示行末的制表符（\t）。</span><br><span class="line">        //表示将匹配到的行末制表符删除。</span><br></pre></td></tr></table></figure>
<p>或者直接使用dos2unix进行转换</p>
<p>处理好的.venn文件可以直接拿来做韦恩图，可以使用在线网站也可以用R，通常情况下我会使用jvenn网站绘制完，根据数据使用AI自己重新绘制一个<br>jvenn网址：<a target="_blank" rel="noopener" href="https://jvenn.toulouse.inrae.fr/app/example.html">https://jvenn.toulouse.inrae.fr/app/example.html</a><br>下面介绍使用R绘制<br>下面是脚本</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line"># 加载必要的库</span><br><span class="line">library(ggVennDiagram)</span><br><span class="line">library(ggplot2)</span><br><span class="line"></span><br><span class="line"># 读取文件数据</span><br><span class="line">file_path &lt;- &quot;Orthogroups.GeneCount.venn&quot;  # 请根据实际文件路径进行替换</span><br><span class="line">data &lt;- read.table(file_path, header = TRUE, sep = &quot;\t&quot;, stringsAsFactors = FALSE)</span><br><span class="line"></span><br><span class="line"># 提取非空值的集合</span><br><span class="line">set_JNYL &lt;- data$JNYL[data$JNYL != &quot;&quot;]</span><br><span class="line">set_WCYL &lt;- data$WCYL[data$WCYL != &quot;&quot;]</span><br><span class="line"></span><br><span class="line"># 创建集合列表</span><br><span class="line">sets &lt;- list(</span><br><span class="line">  JNYL = set_JNYL,</span><br><span class="line">  WCYL = set_WCYL</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"># 使用ggVennDiagram绘制维恩图</span><br><span class="line">ggVennDiagram(sets,</span><br><span class="line">              category.names = c(&quot;JNYL&quot;, &quot;WCYL&quot;), # 集合名字</span><br><span class="line">              set_color = c(&quot;blue&quot;, &quot;red&quot;),       # 集合颜色</span><br><span class="line">              set_size = 6,                       # 集合名字大小</span><br><span class="line">              label = &quot;both&quot;,                     # 标签显示&quot;count&quot;和&quot;percent&quot;</span><br><span class="line">              label_geom = &quot;label&quot;,               # 标签样式</span><br><span class="line">              label_alpha = 0,                    # 标签背景透明</span><br><span class="line">              label_color = &quot;firebrick&quot;,          # 标签颜色</span><br><span class="line">              label_percent_digit = 2,            # 百分比保留2位小数</span><br><span class="line">              edge_lty = &quot;dashed&quot;,                # 边框线型</span><br><span class="line">              edge_size = 1.2                     # 边框粗细</span><br><span class="line">) +</span><br><span class="line">  scale_fill_gradient(low = &quot;grey90&quot;, high = &quot;grey60&quot;) +  # 填充色</span><br><span class="line">  scale_color_manual(values = c(&quot;grey10&quot;, &quot;grey10&quot;))      # 边框颜色</span><br><span class="line"></span><br><span class="line">结果可以使用ggplot2进行优化，这也是为什么要调用ggplot2的原因个人感觉其实到这就可以了</span><br><span class="line">#先将文件转为数据框</span><br><span class="line">venn &lt;- Venn(sets)</span><br><span class="line">df &lt;- process_data(venn)</span><br><span class="line">df</span><br><span class="line">#ggplot2绘图</span><br><span class="line">ggplot()+</span><br><span class="line">  geom_sf(data = venn_region(df), aes(fill=count))+</span><br><span class="line">  geom_sf(data = venn_setedge(df),size=2,lty=&quot;dashed&quot;,color=&quot;grey&quot;)+</span><br><span class="line">  geom_sf(data = venn_setlabel(df),aes(label=name))+</span><br><span class="line">  geom_sf_label(data = venn_region(df),aes(label=id),fontface=&quot;bold&quot;)+</span><br><span class="line">  scale_fill_distiller(palette = 5)+</span><br><span class="line">  theme_void()</span><br></pre></td></tr></table></figure>
<p>其实到这我对图还是不满意，我后面又用AI换了一下颜色</p>
<p><strong>提取独有共有基因的OGid</strong><br>使用jvenn，点击venn图中想要继续分析的区域的数字，即可获得该区域内的所有OGid</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">#将OGid与基因名相匹配</span><br><span class="line">cp ../../Results_May02/Orthogroups/Orthogroups.tsv ./	#将之前结果文件复制到当前目录下</span><br><span class="line">grep -f Ms.txt Orthogroups.tsv | cut -f3 | sed &quot;s/ /\n/g&quot; | sed &quot;s/\t/\n/g&quot; | sed &quot;s/,//g&quot; | sort | uniq &gt; Ms.specific	</span><br><span class="line">#Ms.txt是得到的OGid，必须是unix格式，即使用cat -A打开后结尾是^I^M$，如果不是使用dos2unix转换一下</span><br><span class="line">	-f表示从文件中读取关键词</span><br><span class="line">	| cut -f3 cut 用于按列提取内容，-f3 表示提取每行的第 3 列，因为Ms在Orthogroups.tsv的第三列</span><br><span class="line">&quot;s/ /\n/g&quot;：</span><br><span class="line">    s：表示替换命令。</span><br><span class="line">    / /：表示匹配空格字符。</span><br><span class="line">    \n：表示将空格替换为换行符。</span><br><span class="line">    g：表示全局替换，即替换行中所有的空格</span><br><span class="line">| sed &quot;s/\t/\n/g&quot;：将制表符（\t）替换为换行符</span><br><span class="line">| sed &quot;s/,//g&quot;：将逗号（,）移除</span><br><span class="line">| sort：用于对文本内容进行排序</span><br><span class="line">| uniq： 用于去除相邻的重复行</span><br></pre></td></tr></table></figure>
<p>提取得到基因列表，可以按之前的教程继续后面的GO、KEGG分析</p>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2025/04/18/circos/" rel="prev" title="circos图的绘制">
      <i class="fa fa-chevron-left"></i> circos图的绘制
    </a></div>
      <div class="post-nav-item">
    <a href="/2025/05/19/%E5%88%86%E5%8C%96%E6%97%B6%E9%97%B4%E8%AE%A1%E7%AE%97/" rel="next" title="">
       <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E6%9F%A5%E6%89%BE%E5%85%B1%E6%9C%89%E7%8B%AC%E6%9C%89%E5%9F%BA%E5%9B%A0"><span class="nav-number">1.</span> <span class="nav-text">查找共有独有基因</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-%E5%AE%89%E8%A3%85"><span class="nav-number">1.1.</span> <span class="nav-text">1.安装</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-%E5%87%86%E5%A4%87%E6%95%B0%E6%8D%AE"><span class="nav-number">1.2.</span> <span class="nav-text">2.准备数据</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#3-%E5%AE%B6%E6%97%8F%E8%81%9A%E7%B1%BB%E5%8F%8A%E7%BB%93%E6%9E%9C%E8%A7%A3%E8%AF%BB"><span class="nav-number">1.3.</span> <span class="nav-text">3.家族聚类及结果解读</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#4-%E5%A4%84%E7%90%86%E8%81%9A%E7%B1%BB%E7%BB%93%E6%9E%9C%E5%B9%B6%E7%BB%98%E5%88%B6%E9%9F%A6%E6%81%A9%E5%9B%BE"><span class="nav-number">1.4.</span> <span class="nav-text">4.处理聚类结果并绘制韦恩图</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Plank"
      src="/images/avatar.png">
  <p class="site-author-name" itemprop="name">Plank</p>
  <div class="site-description" itemprop="description">白马？定叫他有来无回！</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">5</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
  </nav>
</div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Plank</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-chart-area"></i>
    </span>
    <span title="站点总字数">54k</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-coffee"></i>
    </span>
    <span title="站点阅读时长">49 分钟</span>
</div>
  <div class="powered-by">由 <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a> 强力驱动
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'Ov23liuVZlZVoSkVGvSX',
      clientSecret: 'fea2d23c05d1fb2cdcd68737c856ed4e1e5dc944',
      repo        : 'MagicWu',
      owner       : 'wefun0454',
      admin       : ['wefun0454'],
      id          : '1c6148f274aedad3d83cc89be1f208f4',
        language: 'zh-CN',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
