<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 7.3.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/1.svg">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/1.svg">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"example.com","root":"/","scheme":"Gemini","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":"gitalk","storage":true,"lazyload":false,"nav":null,"activeClass":"gitalk"},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="分化时间计算分化时间或称分歧时间估计，依据的是分子钟假说，即氨基酸或核苷酸替代速率在进化过程中随时间或进化谱系近似地保持恒定，以某几个特定类群的化石时间作为校正，然后通过基因序列间的分歧程度以及分子钟来估算物种间的分歧时间，同时估算系统发育树上其它节点的发生时间，从而推断相关类群的起源和不同类群的分歧时间。目前，采用依据BI树和ML树估计物种分歧时间的程序很多，例如R8S、MCMCTREE、MUL">
<meta property="og:type" content="article">
<meta property="og:title" content="WeFun blog">
<meta property="og:url" content="http://example.com/2025/05/19/%E5%88%86%E5%8C%96%E6%97%B6%E9%97%B4%E8%AE%A1%E7%AE%97/index.html">
<meta property="og:site_name" content="WeFun blog">
<meta property="og:description" content="分化时间计算分化时间或称分歧时间估计，依据的是分子钟假说，即氨基酸或核苷酸替代速率在进化过程中随时间或进化谱系近似地保持恒定，以某几个特定类群的化石时间作为校正，然后通过基因序列间的分歧程度以及分子钟来估算物种间的分歧时间，同时估算系统发育树上其它节点的发生时间，从而推断相关类群的起源和不同类群的分歧时间。目前，采用依据BI树和ML树估计物种分歧时间的程序很多，例如R8S、MCMCTREE、MUL">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2025-05-19T07:53:23.601Z">
<meta property="article:modified_time" content="2025-05-19T07:27:50.918Z">
<meta property="article:author" content="Plank">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://example.com/2025/05/19/%E5%88%86%E5%8C%96%E6%97%B6%E9%97%B4%E8%AE%A1%E7%AE%97/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'zh-CN'
  };
</script>

  <title> | WeFun blog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">WeFun blog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>归档</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content post posts-expand">
            

    
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://example.com/2025/05/19/%E5%88%86%E5%8C%96%E6%97%B6%E9%97%B4%E8%AE%A1%E7%AE%97/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.png">
      <meta itemprop="name" content="Plank">
      <meta itemprop="description" content="白马？定叫他有来无回！">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="WeFun blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>
              

              <time title="创建时间：2025-05-19 15:53:23 / 修改时间：15:27:50" itemprop="dateCreated datePublished" datetime="2025-05-19T15:53:23+08:00">2025-05-19</time>
            </span>

          <br>
            <span class="post-meta-item" title="本文字数">
              <span class="post-meta-item-icon">
                <i class="far fa-file-word"></i>
              </span>
                <span class="post-meta-item-text">本文字数：</span>
              <span>14k</span>
            </span>
            <span class="post-meta-item" title="阅读时长">
              <span class="post-meta-item-icon">
                <i class="far fa-clock"></i>
              </span>
                <span class="post-meta-item-text">阅读时长 &asymp;</span>
              <span>12 分钟</span>
            </span>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h1 id="分化时间计算"><a href="#分化时间计算" class="headerlink" title="分化时间计算"></a>分化时间计算</h1><pre><code>分化时间或称分歧时间估计，依据的是分子钟假说，即氨基酸或核苷酸替代速率在进化过程中随时间或进化谱系近似地保持恒定，以某几个特定类群的化石时间作为校正，然后通过基因序列间的分歧程度以及分子钟来估算物种间的分歧时间，同时估算系统发育树上其它节点的发生时间，从而推断相关类群的起源和不同类群的分歧时间。目前，采用依据BI树和ML树估计物种分歧时间的程序很多，例如R8S、MCMCTREE、MULTIDIVTIME、BEAST、MEGA等，不同软件通过不同的策略将化石时间信息整合到一个系统发育树中，从而计算得到Divergence time Tree。  
</code></pre>
<p>分子钟假说成立的条件DNA或者蛋白质序列的替代速率是恒定的。20世纪80年代以来，随着DNA序列数据快速积累，大量的证据表明：在长期进化过程中，很多类群的绝大多数基因或蛋白质的序列替换速率根本不符合分子钟假说。对于蛋白质序列，在物种适应辐射过程中，其进化速度可能会大大加快。因此，以蛋白质为基础的恒定进化速率并非理想的分子钟；对于核酸分子，不同基因的分子钟速率不同，但是我们可以认为相同物种某一片段的基因或者蛋白质的替代速率是一定的，在此基础上可以将分子钟模型分为全局分子钟（序列间的期望距离随分歧时间线性增加），局部分子钟模型（不同分支的进化速率不同）<br>在速率恒定的假设下，遗传距离是时间的线性函数，为了将遗传距离转化为分歧时间，至少需要一个能够提供时间信息的标定点（calibration point）。常用的校准信息可以分为：(1)已知的碱基替代速率；(2)化石校准点；(3)生物地理事件校准点；(4)二次校准点<br>本篇教程主要参考：<a target="_blank" rel="noopener" href="https://yanzhongsino.github.io/2021/03/25/bioinfo_phylogeny_caculate.divergence.time/">https://yanzhongsino.github.io/2021/03/25/bioinfo_phylogeny_caculate.divergence.time/</a>  </p>
<h2 id="1-安装软件"><a href="#1-安装软件" class="headerlink" title="1.安装软件"></a>1.安装软件</h2><p>PAML (Phylogenetic Analysis by Maximum Likelihood)，网址：<a target="_blank" rel="noopener" href="https://github.com/abacus-gene/paml">https://github.com/abacus-gene/paml</a><br>官方安装教程网址：<a target="_blank" rel="noopener" href="https://github.com/abacus-gene/paml/wiki/Installation#exporting-paths-linux">https://github.com/abacus-gene/paml/wiki/Installation#exporting-paths-linux</a> ，因此不过多介绍安装流程  </p>
<h2 id="2-准备文件"><a href="#2-准备文件" class="headerlink" title="2.准备文件"></a>2.准备文件</h2><p>多序列比对文件，phylip格式<br>有化石时间标定的进化树（有根树）<br>mcmctree配置文件</p>
<h3 id="2-1检查进化树的结构是否有问题"><a href="#2-1检查进化树的结构是否有问题" class="headerlink" title="2.1检查进化树的结构是否有问题"></a>2.1检查进化树的结构是否有问题</h3><p>ncbi-&gt;taxonomy-&gt;taxonomy common tree（按照物种分类展示的树结构），输入物种的拉丁名，可以看物种的大致进化关系</p>
<h3 id="2-2寻找相应物种的化石时间"><a href="#2-2寻找相应物种的化石时间" class="headerlink" title="2.2寻找相应物种的化石时间"></a>2.2寻找相应物种的化石时间</h3><p>在线网站：<a target="_blank" rel="noopener" href="https://timetree.org/home">https://timetree.org/home</a> ，网站主要记录物种的进化时间和进化关系，会定期更新，有三种搜索模式，1）输入两个物种名称，即可计算得到二者的进化时间，2）提供物种名称，给出该物种的进化事件，3）给出一组物种，他来标定化石时间，一般使用第三种<br>注意查看引用文件，了解使用的是分子时间还是化石时间</p>
<p>展示结果：<br><strong>1）直接输入</strong><br>左侧展示所有研究的分化时间可视化，横坐标文字部分是地质年代，数字为地质年代，曲线从左往右依次是光照，二氧化碳含量，氧含量，陨石撞击事件的陨石坑大小<br>右侧会给出Median Time：XXXMYA（百万年）、大致区间以及所有相关研究给出的大致时间的正态拟合<br>下面表格给出的是相关文献以及对应事件<br><strong>2）提供物种</strong><br>左侧上同<br>右侧展示出物种的进化历史，界门纲目科属种出现的时间<br><strong>3）上传物种名称</strong><br>先显示进化关系，空心的点表示研究相对较少的时间节点，实心点表示研究较多的时间节点，*号表示物种名称被替换成收录名称<br>其他大致一样不过多介绍  </p>
<p>展示出化石时间后，有两种使用方式：<br>1.直接下载相应的树文件，作为有化石时间标定的树文件（左侧最下角，To Newick File）<br>2.copy想要的化石时间，加到自己的树文件中，使用的是RANGE TIME  </p>
<p>一般使用第二种，将之前建树所得的newick树文件中的枝长自展值删去，添加上化石时间范围以及物种数和树的数量（中间用空格隔开），添加之后如下：  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">25 1</span><br><span class="line">((LZ,WX),(JB,(SJ,(YX,(WYZ,(SL,(SD,((DX,(PT,((MGY,NNJ),(YJ,(nigra,CH)))),(MDL,((EL,XZ)&#x27;&gt;1.14&lt;1.30&#x27;,(EZQ,(HGM,((TN,HYM),(XH,(WC,JN))))))))))))))));</span><br><span class="line">((SL,(WYZ,(YX,(SJ,((WX,LZ),JB))))),((((XZ,EL)&#x27;&gt;1.14&lt;1.30&#x27;,(EZQ,(HG,((HYM,TN),(XH,(JN,WC))))))&#x27;&gt;1.3&lt;1.42&#x27;,MDL),(DX,(PT,((NNJ,MGY),(YJ,(CH,nigra)))))),SD); </span><br></pre></td></tr></table></figure>
<p>注意：1代表100MYA，以及之前建树是在ITOL中定的根，一定得是有根树<br>删去自展值可以使用</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sed &#x27;s/:[^,)(]\+//&#x27; Speciestree.txt</span><br></pre></td></tr></table></figure>

<h3 id="2-3获得phylip格式多序列比对文件"><a href="#2-3获得phylip格式多序列比对文件" class="headerlink" title="2.3获得phylip格式多序列比对文件"></a>2.3获得phylip格式多序列比对文件</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cat SpeciesTreeAlignment.fa |tr &#x27;\n&#x27; &#x27;\t&#x27;|sed &#x27;s/&gt;/\n/g&#x27; |sed &#x27;s/\t/      /&#x27;|sed &#x27;s/\t//g&#x27;| awk &#x27;NF &gt; 0&#x27; &gt; supergene.phy.tmp</span><br><span class="line">awk &#x27;&#123;print &quot;  &quot;NR&quot;  &quot;length($2)&#125;&#x27;  supergene.phy.tmp|tail -n 1 | cat -  supergene.phy.tmp &gt;  supergene.phy</span><br></pre></td></tr></table></figure>
<ul>
<li>cat SpeciesTreeAlignment.fa |tr ‘\n’ ‘\t’ (将换行符替换为制表符) |</li>
<li>sed ‘s&#x2F;&gt;&#x2F;\n&#x2F;g’ (将每个序列名前面的&gt;符号替换为换行符) |</li>
<li>sed ‘s&#x2F;\t&#x2F;      &#x2F;‘ (将每行第一个的制表符替换为多个空格) |</li>
<li>sed ‘s&#x2F;\t&#x2F;&#x2F;g’ (删除剩余的制表符,使序列连成一条线) |</li>
<li>awk ‘NF &gt; 0’ (删除空行)&gt; supergene.phy.tmp (临时保存)</li>
<li>awk ‘{print “  “NR”  “length($2)}’  supergene.phy.tmp (计算序列长度)|</li>
<li>tail -n 1 (返回最后一行,包括了行数与序列长度) |</li>
<li>cat -  supergene.phy.tmp (-为上一管道数据) &gt;  supergene.phy(最终文件)（代码来源：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/9f46341d2906">https://www.jianshu.com/p/9f46341d2906</a> ）</li>
</ul>
<p>或者使用软件fast2phy(<a target="_blank" rel="noopener" href="https://github.com/davidmnoriega/fast2phy">https://github.com/davidmnoriega/fast2phy</a> )<br>亦或者使用</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">trimal -in SpeciesTreeAlignment.fa -out supergene.phy -phylip_paml</span><br></pre></td></tr></table></figure>

<p>如果有多个区域的序列，比如exon和intron，LSC、SSC和IR，不同的基因，密码子的第一二三位，需要不同的模型分开估算，那可以把各自区域分别align之后制作多个phy文件，再合并到一起，用空行隔开，组成input.phy文件。（此时mcmctree.ctl的ndata值为区域的个数）  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#OG.length包含两列，一列OGID，一列比对后序列长度</span><br><span class="line">cat ./OG.length | while read line</span><br><span class="line">do</span><br><span class="line">        sample_id=$(echo $line |awk &#x27;&#123;print $1&#125;&#x27;) #获取OG.ID</span><br><span class="line">        sample_a=$(echo $line |awk &#x27;&#123;print $2&#125;&#x27;) #获取序列长度</span><br><span class="line">        sed &quot;s/_.*//g&quot; ../singlegenetree/$&#123;sample_id&#125;/$&#123;sample_id&#125;.mafft.pep|seqkit seq -w 0|sed -E &quot;:a;N;s/\n/ /g;ta&quot; |sed &quot;s/ &gt;/\n/g&quot; |sed &quot;s/&gt;//g&quot;|sed &quot;s/ /  /g&quot;|sed &quot;1i\5  $&#123;sample_a&#125;&quot; |sed &#x27;1i\ &#x27; &gt; ./phy/$&#123;sample_id&#125;.phy #把上一步获取的$&#123;sample_id&#125;.mafft.pep改为phylip格式，并在首行添加空格行（为了合并后每个OG用空行隔开）</span><br><span class="line"></span><br><span class="line">done</span><br><span class="line">cat ./phy/*phy &gt;input.phy #合并所有phy文件为一个.phy</span><br></pre></td></tr></table></figure>
<p>（代码来源：<a target="_blank" rel="noopener" href="https://yanzhongsino.github.io/2021/03/25/bioinfo_phylogeny_caculate.divergence.time/">https://yanzhongsino.github.io/2021/03/25/bioinfo_phylogeny_caculate.divergence.time/</a> ）</p>
<h2 id="2-4mcmctree配置文件"><a href="#2-4mcmctree配置文件" class="headerlink" title="2.4mcmctree配置文件"></a>2.4mcmctree配置文件</h2><p>mcmctree配置文件为mcmctree.ctl，具体参数及介绍如下</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br></pre></td><td class="code"><pre><span class="line">seed = -1 *设置随机数作为seed，-1代表使用系统当前时间作为随机数</span><br><span class="line">seqfile = supergene.phy *输入多序列比对文件</span><br><span class="line">treefile =MCMTREE.tree *带校准点（化石时间）的有根树文件</span><br><span class="line">outfile = mcmc.out *输出文件</span><br><span class="line">mcmcfile = mcmc.txt *输出的mcmc信息文件，可用Tracer软件查看</span><br><span class="line"></span><br><span class="line">seqtype = 0  * 设置多序列比对数据类型；0：核酸数据；1：密码子比对数据；2：氨基酸数据；</span><br><span class="line">usedata = 3</span><br><span class="line">* 是否利用多序列比对数据；</span><br><span class="line">* 0: no data不使用，不会进行likelihood估算，会快速得到mcmc树，但分歧时间不可用; </span><br><span class="line">* 1:seq like，使用多序列比对数据进行likelihood估算，正常进行mcmc; usedata=1时model无法选择；</span><br><span class="line">* 2:normal 进行正常的approximation likelihood分析，不读取多序列比对数据，直接读取当前目录的in.BV文件，in.BV是由usedata = 3时生成的out.BV重命名得来；此外，由于程序BUG，当设置usedata = 2时，一定要在改行参数后加 *，否则程序报错 Error: file name empty..；</span><br><span class="line">* 3：程序利用多序列比对数据调用baseml/codeml命令对数据进行分析，生成out.BV文件。由于mcmctree调用baseml/codeml进行估算的参数设置可能不太好（特别时对蛋白序列进行估算时），推荐自己修改软件自动生成的baseml/codeml配置文件，然后再手动运行baseml/codeml命令，再整合其结果文件为out.BV文件。</span><br><span class="line"></span><br><span class="line">ndata = 1    * 输入的多序列比对的数据区域的数量；</span><br><span class="line">clock = 2    * 设置分子钟算法，1: global clock，表示所有分支进化速率一致; 2: independent rates，各分支的进化速率独立且进化速率的对数log(r)符合正态分布; 3，correlated rates方法，和方法2类似，但是log(r)的方差和时间t相关。</span><br><span class="line">*       TipDate = 1 100  *当外部节点由取样时间时使用该参数进行设置，同时该参数也设置了时间单位。具体数据示例请见examples/TipData文件夹。</span><br><span class="line">RootAge = &lt;10  * constraint on root age, used if no fossil for root.设置root节点的分歧时间，一般设置一个最大值。</span><br><span class="line"></span><br><span class="line">model = 0    * models for DNA:</span><br><span class="line">                        * 0:JC69, 1:K80, 2:F81, 3:F84, 4:HKY85；*设置碱基替换模型；当设置usedata = 1时，model不能使用超过4的模型，所以usedata = 1时用model = 4</span><br><span class="line">                      * models for codons:</span><br><span class="line">                        * 0:one 恒定速率模型, 1:b 中性模型, 2:2 or more dN/dS ratios for branches 选择模型。</span><br><span class="line">                      * models for AAs or codon-translated AAs:</span><br><span class="line">			* 0：Poisson 模型</span><br><span class="line">			* 1：Proportional 模型</span><br><span class="line">			*2：经验模型</span><br><span class="line">			*3：经验模型 + F</span><br><span class="line">			*6：基于密码子模型</span><br><span class="line">			*8：REVaa_0 模型</span><br><span class="line">			*9：REVaa (nr=189) 模型</span><br><span class="line">alpha = 0.5   * alpha for gamma rates at sites；*核酸序列中不同位点，其进化速率不一致，其变异速率服从GAMMA分布。一般设置GAMMA分布的alpha值为0.5。若该参数值设置为0，则表示所有位点的进化速率一致。此外，当userdata = 2时，alpha、ncatG、alpha_gamma、kappa_gamma这4个GAMMA参数无效。因为userdata = 2时，不会利用到多序列比对的数据。model=4时一般设为0.5，否则一般设为0</span><br><span class="line">ncatG = 5    * No. categories in discrete gamma；设置离散型GAMMA分布的categories值。</span><br><span class="line"></span><br><span class="line">cleandata = 0    * remove sites with ambiguity data (1:yes, 0:no)?是否去除gap</span><br><span class="line"></span><br><span class="line">BDparas = 1 1 0.1   * birth, death, sampling；*设置出生率、死亡率和取样比例。若输入有根树文件中的时间单位发生改变，则需要相应修改出生率和死亡率的值。例如，时间单位由100Myr变换为1Myr，则要设置成&quot;.01 .01 0.1&quot;。</span><br><span class="line">kappa_gamma = 6 2      * gamma prior for kappa；设置kappa（转换/颠换比率）的GAMMA分布参数。</span><br><span class="line">alpha_gamma = 1 1      * gamma prior for alpha；设置GAMMA形状参数alpha的GAMMA分布参数。</span><br><span class="line"></span><br><span class="line">rgene_gamma = 2 20 1   * gamma prior for rate for genes；设置序列中所所有位点平均[碱基/密码子/氨基酸]替换率的Dirichlet-GAMMA分布参数：alpha=2、beta=20、初始平均替换率为每100million年（取决于输入有根树文件中的时间单位）1个替换。若时间单位由100Myr变换为1Myr，则要设置成&quot;2 2000 1&quot;。总体上的平均进化速率为：2 / 20 = 0.1 个替换 / 每100Myr，即每个位点每年的替换数为 1e-9。</span><br><span class="line">sigma2_gamma = 1 10 1    * gamma prior for sigma^2     (for clock=2 or 3)；设置所有位点进化速率取对数后方差（sigma的平方）的Dirichlet-GAMMA分布参数：alpha=1、beta=10、初始方差值为1。当clock参数值为1时，表示使用全局的进化速率，各分枝的进化速率没有差异，即方差为0，该参数无效；当clock参数值为2时，若修改了时间单位，该参数值不需要改变；当clock参数值为3时，若修改了时间单位，该参数值需要改变。</span><br><span class="line"></span><br><span class="line">finetune = 1: .1 .1 .1 .1 .1 .1  * times, rates, mixing, paras, RateParas；冒号前的值设置是否自动进行finetune，一般设置成1，然程序自动进行优化分析；冒号后面设置各个参数的步进值：times, musigma2, rates, mixing, paras, FossilErr。由于有了自动设置，该参数不像以前版本那么重要了，可能以后会取消该参数。</span><br><span class="line"></span><br><span class="line">print = 1  *设置打印mcmc的取样信息：0，不打印mcmc结果；1，打印除了分支进化速率的其它信息（各内部节点分歧时间、平均进化速率、sigma2值）；2，打印所有信息。 </span><br><span class="line">burnin = 1000000  *将前1000000次迭代burnin后，再进行取样（即打印出该次迭代估算的结果信息，各内部节点分歧时间、平均进化速率、sigma2值和各分支进化速率等）。</span><br><span class="line">sampfreq = 10  *每10次迭代则取样一次</span><br><span class="line">nsample = 500000  *当取样次数达到该次数时，则取样结束，同时结束程序。</span><br><span class="line"></span><br><span class="line">*** Note: Make your window wider (100 columns) when running this program.</span><br></pre></td></tr></table></figure>

<h2 id="3-运行"><a href="#3-运行" class="headerlink" title="3.运行"></a>3.运行</h2><p>完成后运行<br>mcmctree mcmctree.ctl</p>
<h2 id="4-其他方法"><a href="#4-其他方法" class="headerlink" title="4.其他方法"></a>4.其他方法</h2><h3 id="4-1氨基酸序列计算分化时间"><a href="#4-1氨基酸序列计算分化时间" class="headerlink" title="4.1氨基酸序列计算分化时间"></a>4.1氨基酸序列计算分化时间</h3><p>一般不建议使用氨基酸序列去计算分化时间，因为一般需要计算氨基酸替换速率，并选择相应的模型，而使用核酸序列相对简单，但是如果想要直接使用之前Orthofinder的结果，也就是氨基酸序列进行分化时间的计算，则需要进行一些处理  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">#按照上面的处理之后先计算氨基酸替换速率，修改tmp0001.ctl的内容</span><br><span class="line">cp tmp0001.ctl codeml.ctl</span><br><span class="line">#添加参数，增加gamma分布的先验数据</span><br><span class="line">echo &quot;clock =1&quot; &gt;&gt;codeml.ctl</span><br><span class="line"></span><br><span class="line">#然后修改一些参数</span><br><span class="line">model = 2 * model=0使用的是简单的一致模型计算氨基酸替换速率，一般建议使用2，使用WAG模型进行计算</span><br><span class="line">aaRatefile = /home/walnut168/mnt/wf/program/paml-4.10.7/dat/wag.dat * paml目录下的文件一般在paml/dat/wag.dat</span><br></pre></td></tr></table></figure>
<p>修改树文件<br>将tmp0001.trees文件改为有根树，并添加上化石标定时间，需要添加具体时间，一般取范围的中间数就行<br>我的原文件为  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">  1</span><br><span class="line"></span><br><span class="line">(LZ, WX, (JB, (SJ, (YX, (WYZ, (SL, (SD, ((DX, (PT, ((MGY, NNJ), (YJ, (nigra, CH))))), (MDL, ((EL, XZ), (EZQ, (HG, ((TN, HYM), (XH, (WC, JN)))))))))))))));</span><br></pre></td></tr></table></figure>
<p>修改后为</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">  1</span><br><span class="line"></span><br><span class="line">((LZ, WX), (JB, (SJ, (YX, (WYZ, (SL, (SD, ((DX, (PT, ((MGY, NNJ), (YJ, (nigra, CH))))), (MDL, ((EL, XZ)&#x27;@1.22&#x27;, (EZQ, (HG, ((TN, HYM), (XH, (WC, JN)))))))))))))));</span><br></pre></td></tr></table></figure>
<p>修改完成后运行</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">codeml codeml.ctl</span><br></pre></td></tr></table></figure>
<p>就算完成查看tmp0001.out，替换速率在“Substitution rate is per time unit” 下一行<br>完成后修改mcmtree.ctl文件中的rgene_gamma &#x3D; 2 20，根据氨基酸替换速率修改，例如我的氨基酸替换速率为0.31，则大致修改为rgene_gamma &#x3D; 2 6<br>计算完成之后重新运行mcmctree mcmctree.ctl，重新生成out.BV文件<br>后续操作等于重新进行一次分化时间的计算，最后获得目标文件  </p>
<p>注意，运行时，程序运行日志中有进度，5%-100%后面第一列数字为模型接受程度，一般30%左右较好，最大波动为15%-70%，如果超过，最好修改burnin参数，并适当增加迭代次数等等，第二列为预估的分化时间，正常情况下应该是不断稳定下来，如果靠近100%处还波动较大，应该增加迭代次数，增加nsample，增加sampfreq  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"> 5% 0.42 0.29 0.45 0.40 0.25  9.662 7.289 9.250 8.390 7.561 6.124 - 0.067 0.902 -23.42  0:02</span><br><span class="line"> 10% 0.38 0.30 0.36 0.41 0.24  9.856 7.443 9.520 8.836 7.780 6.542 - 0.064 0.831 -23.58  0:03</span><br><span class="line"> 15% 0.40 0.30 0.39 0.41 0.19  9.811 7.373 9.498 8.704 7.693 6.664 - 0.063 0.812 -23.57  0:03</span><br><span class="line"> 20% 0.40 0.30 0.40 0.41 0.17  9.807 7.439 9.465 8.616 7.630 6.722 - 0.063 0.786 -23.50  0:04</span><br><span class="line"> 25% 0.38 0.29 0.40 0.40 0.15  9.754 7.513 9.436 8.632 7.685 6.842 - 0.063 0.753 -23.49  0:05</span><br><span class="line"> 30% 0.38 0.30 0.40 0.39 0.15  9.778 7.412 9.467 8.672 7.747 6.909 - 0.063 0.722 -23.49  0:05</span><br><span class="line"> 35% 0.37 0.30 0.39 0.39 0.16  9.800 7.532 9.493 8.722 7.812 6.939 - 0.064 0.696 -23.53  0:06</span><br><span class="line"> 40% 0.37 0.30 0.39 0.39 0.16  9.835 7.668 9.543 8.770 7.866 6.919 - 0.064 0.677 -23.53  0:06</span><br><span class="line"> 45% 0.35 0.29 0.41 0.41 0.16  9.853 7.787 9.546 8.745 7.809 6.883 - 0.064 0.659 -23.49  0:07</span><br><span class="line"> 50% 0.34 0.29 0.40 0.42 0.16  9.811 7.843 9.483 8.703 7.756 6.855 - 0.064 0.644 -23.51  0:08</span><br><span class="line"> 55% 0.34 0.29 0.40 0.41 0.16  9.800 7.875 9.478 8.706 7.783 6.884 - 0.065 0.633 -23.51  0:08</span><br><span class="line"> 60% 0.33 0.29 0.40 0.41 0.16  9.829 7.960 9.509 8.740 7.808 6.918 - 0.064 0.621 -23.52  0:09</span><br><span class="line"> 65% 0.33 0.29 0.40 0.42 0.15  9.843 8.032 9.526 8.760 7.800 6.944 - 0.064 0.614 -23.52  0:09</span><br><span class="line"> 70% 0.33 0.28 0.40 0.42 0.15  9.852 8.055 9.540 8.775 7.824 6.978 - 0.064 0.609 -23.52  0:10</span><br><span class="line"> 75% 0.32 0.28 0.40 0.42 0.15  9.866 8.113 9.560 8.789 7.842 6.991 - 0.064 0.604 -23.53  0:10</span><br><span class="line"> 80% 0.32 0.28 0.41 0.42 0.15  9.886 8.172 9.576 8.803 7.832 6.993 - 0.064 0.596 -23.54  0:11</span><br><span class="line"> 85% 0.31 0.28 0.40 0.43 0.15  9.891 8.207 9.587 8.825 7.832 6.981 - 0.064 0.586 -23.54  0:12</span><br><span class="line"> 90% 0.31 0.28 0.40 0.43 0.16  9.907 8.214 9.609 8.849 7.829 6.972 - 0.064 0.580 -23.55  0:12</span><br><span class="line"> 95% 0.32 0.28 0.40 0.43 0.16  9.919 8.234 9.621 8.856 7.843 6.968 - 0.064 0.572 -23.56  0:13</span><br><span class="line">100% 0.32 0.28 0.41 0.43 0.16  9.924 8.218 9.627 8.854 7.837 6.950 - 0.064 0.565 -23.55  0:13</span><br></pre></td></tr></table></figure>


<p>除此之外还需要观察mcmc.out文件最后部分，观察t_n部分，该部分为内部节点的分化时间，观察一下是否波动不大  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">Posterior mean (95% Equal-tail CI) (95% HPD CI) HPD-CI-width</span><br><span class="line"></span><br><span class="line">t_n26          9.7170 ( 8.0432, 10.8710) ( 8.0329, 10.8514)  2.8185 (Jnode 48)</span><br><span class="line">t_n27          8.8041 ( 6.8167, 10.2398) ( 6.8624, 10.2803)  3.4178 (Jnode 47)</span><br><span class="line">t_n28          9.4250 ( 7.7433, 10.6173) ( 7.8992, 10.7086)  2.8094 (Jnode 46)</span><br><span class="line">t_n29          8.6206 ( 6.9800,  9.8600) ( 7.1669, 10.0149)  2.8481 (Jnode 45)</span><br><span class="line">t_n30          7.3818 ( 5.9074,  8.4911) ( 5.9742,  8.5341)  2.5599 (Jnode 44)</span><br><span class="line">t_n31          6.1850 ( 4.7974,  7.4212) ( 4.8173,  7.4297)  2.6125 (Jnode 43)</span><br><span class="line">t_n32          5.6159 ( 4.2569,  6.9734) ( 4.2563,  6.9714)  2.7152 (Jnode 42)</span><br><span class="line">t_n33          4.8949 ( 3.7075,  6.2810) ( 3.5774,  6.0312)  2.4538 (Jnode 41)</span><br><span class="line">t_n34          4.3230 ( 3.1748,  5.2996) ( 3.2144,  5.3294)  2.1149 (Jnode 40)</span><br><span class="line">t_n35          3.1460 ( 2.3840,  3.8388) ( 2.3812,  3.8352)  1.4540 (Jnode 39)</span><br><span class="line">t_n36          2.5618 ( 1.9859,  3.1193) ( 1.9480,  3.0724)  1.1243 (Jnode 38)</span><br><span class="line">t_n37          2.1956 ( 1.6387,  2.6295) ( 1.6494,  2.6355)  0.9861 (Jnode 37)</span><br><span class="line">t_n38          1.6816 ( 0.9820,  2.2494) ( 1.0505,  2.2811)  1.2306 (Jnode 36)</span><br><span class="line">t_n39          1.7495 ( 1.2646,  2.2873) ( 1.2470,  2.2652)  1.0182 (Jnode 35)</span><br><span class="line">t_n40          0.1633 ( 0.0571,  0.3442) ( 0.0426,  0.2922)  0.2496 (Jnode 34)</span><br><span class="line">t_n41          3.7507 ( 2.8133,  4.5681) ( 2.8132,  4.5679)  1.7547 (Jnode 33)</span><br><span class="line">t_n42          2.9491 ( 2.2476,  3.6595) ( 2.2476,  3.6593)  1.4117 (Jnode 32)</span><br><span class="line">t_n43          1.1929 ( 1.1358,  1.2913) ( 1.1322,  1.2855)  0.1533 (Jnode 31)</span><br><span class="line">t_n44          0.7227 ( 0.4832,  1.0478) ( 0.4510,  0.9966)  0.5456 (Jnode 30)</span><br><span class="line">t_n45          0.3438 ( 0.2135,  0.4484) ( 0.2025,  0.4341)  0.2315 (Jnode 29)</span><br><span class="line">t_n46          0.3157 ( 0.1971,  0.4061) ( 0.1916,  0.3997)  0.2082 (Jnode 28)</span><br><span class="line">t_n47          0.2863 ( 0.1536,  0.3715) ( 0.1659,  0.3797)  0.2139 (Jnode 27)</span><br><span class="line">t_n48          0.1475 ( 0.0932,  0.2344) ( 0.0897,  0.2286)  0.1389 (Jnode 26)</span><br><span class="line">t_n49          0.1210 ( 0.0707,  0.2043) ( 0.0671,  0.1954)  0.1283 (Jnode 25)</span><br><span class="line">mu             0.0954 ( 0.0759,  0.1246) ( 0.0744,  0.1216)  0.0472</span><br><span class="line">sigma2         0.2907 ( 0.1821,  0.4506) ( 0.1740,  0.4368)  0.2627</span><br><span class="line">lnL          -23.5714 (-34.0770, -15.1300) (-33.1800, -14.4370) 18.7430</span><br></pre></td></tr></table></figure>
<p>一般近似似然法计算就是会有较大的误差，所以还是推荐使用usedata &#x3D; 1</p>
<h2 id="4-2-4D位点计算分化时间"><a href="#4-2-4D位点计算分化时间" class="headerlink" title="4.2 4D位点计算分化时间"></a>4.2 4D位点计算分化时间</h2><p>什么是4D位点？一个遗传密码子通常由三个核苷酸构成，从左到右依次为第一个位点、第二个位点、第三个位点。如果密码子的某个位点上无论是哪种核苷酸，均编码同样的氨基酸，则称这个位点为 4 倍简并位点，这个位点一般是密码子的第三位。4DTV( four-fold synonymous (degenerative) third-codon transversion) 表示4D位点（fourfold degenerate site）上发生颠换（嘌呤突变为嘧啶或者嘧啶突变为嘌呤）的位点替换率。</p>
<p>使用单拷贝基因家族的4D位点计算分化时间，则还需要准备以下文件</p>
<ul>
<li>1）CDS文件</li>
<li>2）单拷贝基因家族列表<br>Orthogroups_SingleCopyOrthologues.txt（基因家族ID）</li>
<li>3）单拷贝基因家族的序列</li>
<li>4）基因家族统计列表<br>Orthogroups.tsv（基因ID）</li>
</ul>
<p>首先根据单拷贝基因家族列表提取出基因id  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">less Orthogroups_SingleCopyOrthologues.txt | while read aa ;do grep $aa Orthogroups.tsv ;done &gt; id.txt</span><br></pre></td></tr></table></figure>
<p>或者</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">awk &#x27;&#123;if(NR==FNR)&#123;A[$1]=1&#125;else&#123; if($1 in A || FNR==1)&#123;print $0&#125;&#125; &#125;&#x27; \</span><br><span class="line">Orthogroups_SingleCopyOrthologues.txt Orthogroups.tsv \</span><br><span class="line">&gt; single_copy.txt</span><br></pre></td></tr></table></figure>
<p>或者</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk &#x27;&#123;if(NR==1)&#123;f=NF&#125;; if( $0 !~ /,/ &amp;&amp; NF==f )&#123;print $0&#125;&#125;&#x27; Orthogroups.tsv &gt; single_copy.txt</span><br></pre></td></tr></table></figure>
<p>在每个ID前加上物种前缀</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">awk &#x27;&#123;if(NR==1)&#123; n=split($0,A, &quot;\t&quot;)&#125; else &#123; for(i=2; i&lt;= n; i++)&#123; \</span><br><span class="line">printf A[i]&quot;_&quot;$i&quot;\t&quot; &#125;; printf &quot;\n&quot; &#125; &#125;&#x27; single_copy.txt | \</span><br><span class="line">sed &#x27;s/\s\+$//&#x27; &gt; single_copy.txt.change</span><br></pre></td></tr></table></figure>
<p>给每一个cds序列添加上物种标签</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">for i in cds/*.fa;do</span><br><span class="line">base=$(basename &quot;$i&quot; .fa);</span><br><span class="line">sed &quot;s/^&gt;/&gt;$&#123;base&#125;_/&quot; &quot;$i&quot; &gt;&quot;$base.fa.change&quot;;</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<p>protein序列进行类似操作</p>
<p>合并cds和protein文件</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cat changecds/*.fa.change &gt; all.cds.fa</span><br><span class="line">cat changepep/*.fa.change &gt;all.pep.fa</span><br></pre></td></tr></table></figure>

<p>使用ParaAT（<a target="_blank" rel="noopener" href="https://github.com/wonaya/ParaAT">https://github.com/wonaya/ParaAT</a> ）提取相应的cds文件  </p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">echo &quot;6&quot; &gt; proc.txt</span><br><span class="line">ParaAT.pl -h single_copy.txt.change -a all.pep.fa -n all.cds.fa -o Para_out -p proc.txt</span><br><span class="line"></span><br><span class="line">-h single_copy.txt \ #基因家族信息，每一行一个家族</span><br><span class="line">-a all.pep.fa \ #全部蛋白序列</span><br><span class="line">-n all.cds.fa \ #全部cds序列</span><br><span class="line">-o Para_out \ #输出结果目录</span><br><span class="line">-p proc.txt #指定线程数文件</span><br></pre></td></tr></table></figure>
<p>注意protein文件和cds文件名称必须与single_copy.txt.change一致，为保证这一点可以进行以下操作</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">#简化序列名</span><br><span class="line">sed -i &#x27;/^&gt;/ s/ .*//&#x27; changepep/*.fa.change</span><br><span class="line">#直接将cds翻译成protein保证名称一致</span><br><span class="line">for i in cds/*.fa ;do</span><br><span class="line">base=$(basename &quot;$i&quot; .fa);</span><br><span class="line">seqkit translate &quot;$i&quot; &gt; pep/&quot;$base.fa&quot;;</span><br><span class="line">done</span><br></pre></td></tr></table></figure>
<p>除此之外，序列名称不允许包含|，：，-等特殊字符，并且序列数也有限制，因此只推荐数量较少的进行</p>
<p>序列数较多可以使用Hyphy（<a target="_blank" rel="noopener" href="https://github.com/veg/hyphy-analyses/tree/master/codon-msa">https://github.com/veg/hyphy-analyses/tree/master/codon-msa</a> ）具体流程较为复杂，需要重新编写脚本，不推荐使用</p>
<p>到此分化时间估计就完成了，可视化也是使用ITOL网站，但是由于文件格式原因，不能直接像之前的树文件一样拖入即可，因此需要进行一些修改，此部分在下一篇教程《基因家族扩张收缩》中介绍</p>

    </div>

    
    
    

      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/2025/05/19/%E6%9F%A5%E6%89%BE%E5%85%B1%E6%9C%89%E5%92%8C%E7%8B%AC%E6%9C%89%E5%9F%BA%E5%9B%A0/" rel="prev" title="">
      <i class="fa fa-chevron-left"></i> 
    </a></div>
      <div class="post-nav-item">
    <a href="/2025/05/19/%E5%9F%BA%E5%9B%A0%E5%AE%B6%E6%97%8F%E6%89%A9%E5%BC%A0%E6%94%B6%E7%BC%A9/" rel="next" title="">
       <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  



          </div>
          
    <div class="comments" id="gitalk-container"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%88%86%E5%8C%96%E6%97%B6%E9%97%B4%E8%AE%A1%E7%AE%97"><span class="nav-number">1.</span> <span class="nav-text">分化时间计算</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#1-%E5%AE%89%E8%A3%85%E8%BD%AF%E4%BB%B6"><span class="nav-number">1.1.</span> <span class="nav-text">1.安装软件</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-%E5%87%86%E5%A4%87%E6%96%87%E4%BB%B6"><span class="nav-number">1.2.</span> <span class="nav-text">2.准备文件</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#2-1%E6%A3%80%E6%9F%A5%E8%BF%9B%E5%8C%96%E6%A0%91%E7%9A%84%E7%BB%93%E6%9E%84%E6%98%AF%E5%90%A6%E6%9C%89%E9%97%AE%E9%A2%98"><span class="nav-number">1.2.1.</span> <span class="nav-text">2.1检查进化树的结构是否有问题</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-2%E5%AF%BB%E6%89%BE%E7%9B%B8%E5%BA%94%E7%89%A9%E7%A7%8D%E7%9A%84%E5%8C%96%E7%9F%B3%E6%97%B6%E9%97%B4"><span class="nav-number">1.2.2.</span> <span class="nav-text">2.2寻找相应物种的化石时间</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-3%E8%8E%B7%E5%BE%97phylip%E6%A0%BC%E5%BC%8F%E5%A4%9A%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9%E6%96%87%E4%BB%B6"><span class="nav-number">1.2.3.</span> <span class="nav-text">2.3获得phylip格式多序列比对文件</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#2-4mcmctree%E9%85%8D%E7%BD%AE%E6%96%87%E4%BB%B6"><span class="nav-number">1.3.</span> <span class="nav-text">2.4mcmctree配置文件</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#3-%E8%BF%90%E8%A1%8C"><span class="nav-number">1.4.</span> <span class="nav-text">3.运行</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#4-%E5%85%B6%E4%BB%96%E6%96%B9%E6%B3%95"><span class="nav-number">1.5.</span> <span class="nav-text">4.其他方法</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#4-1%E6%B0%A8%E5%9F%BA%E9%85%B8%E5%BA%8F%E5%88%97%E8%AE%A1%E7%AE%97%E5%88%86%E5%8C%96%E6%97%B6%E9%97%B4"><span class="nav-number">1.5.1.</span> <span class="nav-text">4.1氨基酸序列计算分化时间</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#4-2-4D%E4%BD%8D%E7%82%B9%E8%AE%A1%E7%AE%97%E5%88%86%E5%8C%96%E6%97%B6%E9%97%B4"><span class="nav-number">1.6.</span> <span class="nav-text">4.2 4D位点计算分化时间</span></a></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Plank"
      src="/images/avatar.png">
  <p class="site-author-name" itemprop="name">Plank</p>
  <div class="site-description" itemprop="description">白马？定叫他有来无回！</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">5</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
  </nav>
</div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2025</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Plank</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-chart-area"></i>
    </span>
    <span title="站点总字数">54k</span>
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-coffee"></i>
    </span>
    <span title="站点阅读时长">49 分钟</span>
</div>
  <div class="powered-by">由 <a href="https://hexo.io/" class="theme-link" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.org/" class="theme-link" rel="noopener" target="_blank">NexT.Gemini</a> 强力驱动
  </div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

<link rel="stylesheet" href="//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.css">

<script>
NexT.utils.loadComments(document.querySelector('#gitalk-container'), () => {
  NexT.utils.getScript('//cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js', () => {
    var gitalk = new Gitalk({
      clientID    : 'Ov23liuVZlZVoSkVGvSX',
      clientSecret: 'fea2d23c05d1fb2cdcd68737c856ed4e1e5dc944',
      repo        : 'MagicWu',
      owner       : 'wefun0454',
      admin       : ['wefun0454'],
      id          : '3c5db5203c57657a3a2feb426cf57466',
        language: 'zh-CN',
      distractionFreeMode: true
    });
    gitalk.render('gitalk-container');
  }, window.Gitalk);
});
</script>

</body>
</html>
